// styl3/dist/decorators/decorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.symbolsFunctions = exports.decoratorMap = exports.decoratorFunctions = exports.symbolResets = exports.symbolMap = void 0;
const utils_1 = require("../utils/utils");
const decoratorMap = {
    bold: '*',
    dim: '~',
    italic: '%',
    underline: '!',
    blink: '^',
    invert: '@',
    hidden: '#',
    strikeout: '$'
};
exports.decoratorMap = decoratorMap;
exports.symbolMap = {
    bold: 1,
    dim: 2,
    italic: 3,
    underline: 4,
    blink: 5,
    invert: 7,
    hidden: 8,
    strikeout: 9
};
exports.symbolResets = {
    bold: 22,
    dim: 22,
    italic: 23,
    underline: 24,
    blink: 25,
    invert: 27,
    hidden: 28,
    strikeout: 29
};
const symbolsFunctions = {
    bold: (color) => (0, utils_1.replaceColor)('bold', color),
    dim: (color) => (0, utils_1.replaceColor)('dim', color),
    italic: (color) => (0, utils_1.replaceColor)('italic', color),
    underline: (color) => (0, utils_1.replaceColor)('underline', color),
    blink: (color) => (0, utils_1.replaceColor)('blink', color),
    invert: (color) => (0, utils_1.replaceColor)('invert', color),
    hidden: (color) => (0, utils_1.replaceColor)('hidden', color),
    strikeout: (color) => (0, utils_1.replaceColor)('strikeout', color),
};
exports.symbolsFunctions = symbolsFunctions;
const decoratorFunctions = {
    bold: (txt, ...args) => symbolsFunctions.bold((0, utils_1.getValue)(txt, ...args)),
    dim: (txt, ...args) => symbolsFunctions.dim((0, utils_1.getValue)(txt, ...args)),
    italic: (txt, ...args) => symbolsFunctions.italic((0, utils_1.getValue)(txt, ...args)),
    underline: (txt, ...args) => symbolsFunctions.underline((0, utils_1.getValue)(txt, ...args)),
    blink: (txt, ...args) => symbolsFunctions.blink((0, utils_1.getValue)(txt, ...args)),
    invert: (txt, ...args) => symbolsFunctions.invert((0, utils_1.getValue)(txt, ...args)),
    hidden: (txt, ...args) => symbolsFunctions.hidden((0, utils_1.getValue)(txt, ...args)),
    strikeout: (txt, ...args) => symbolsFunctions.strikeout((0, utils_1.getValue)(txt, ...args))
};
exports.decoratorFunctions = decoratorFunctions;
