// chartscii/dist/index.js
// formatters/formatter.ts
import style from "styl3";

// utils/color.ts
var NAMED_COLORS = {
  red: [255, 0, 0],
  green: [0, 255, 0],
  blue: [0, 0, 255],
  yellow: [255, 255, 0],
  cyan: [0, 255, 255],
  magenta: [255, 0, 255],
  purple: [128, 0, 128],
  orange: [255, 165, 0],
  pink: [255, 192, 203],
  white: [255, 255, 255],
  black: [0, 0, 0],
  gray: [128, 128, 128],
  grey: [128, 128, 128],
  marine: [0, 119, 190],
  lime: [0, 255, 0],
  teal: [0, 128, 128],
  navy: [0, 0, 128],
  maroon: [128, 0, 0],
  olive: [128, 128, 0],
  aqua: [0, 255, 255],
  coral: [255, 127, 80],
  salmon: [250, 128, 114],
  gold: [255, 215, 0],
  violet: [238, 130, 238],
  indigo: [75, 0, 130],
  turquoise: [64, 224, 208],
  crimson: [220, 20, 60],
  chocolate: [210, 105, 30],
  tomato: [255, 99, 71],
  skyblue: [135, 206, 235]
};
function isGradientObject(color) {
  return typeof color === "object" && color !== null && color.type === "gradient" && Array.isArray(color.colors);
}
function isGradientString(color) {
  return typeof color === "string" && color.startsWith("gradient(") && color.endsWith(")");
}
function parseGradient(value) {
  if (!value.startsWith("gradient(") || !value.endsWith(")")) {
    return value;
  }
  const content = value.slice(9, -1);
  if (!content) {
    return value;
  }
  const parts = content.split(":");
  const colorsPart = parts[0];
  const options = parts.slice(1);
  const colors = colorsPart.split(",").map((c) => c.trim()).filter((c) => c.length > 0);
  if (colors.length === 0) {
    return value;
  }
  const gradient = {
    type: "gradient",
    colors
  };
  for (const opt of options) {
    const trimmed = opt.trim().toLowerCase();
    if (trimmed === "horizontal" || trimmed === "vertical" || trimmed === "diagonal") {
      gradient.direction = trimmed;
    } else if (trimmed === "reverse" || trimmed === "reversed") {
      gradient.reverse = true;
    }
  }
  return gradient;
}
function normalizeColor(color) {
  if (typeof color === "string" && isGradientString(color)) {
    return parseGradient(color);
  }
  return color;
}
function ansi256ToRgb(index) {
  if (index < 16) {
    const standard = [
      [0, 0, 0],
      // 0: black
      [128, 0, 0],
      // 1: red
      [0, 128, 0],
      // 2: green
      [128, 128, 0],
      // 3: yellow
      [0, 0, 128],
      // 4: blue
      [128, 0, 128],
      // 5: magenta
      [0, 128, 128],
      // 6: cyan
      [192, 192, 192],
      // 7: white
      [128, 128, 128],
      // 8: bright black (gray)
      [255, 0, 0],
      // 9: bright red
      [0, 255, 0],
      // 10: bright green
      [255, 255, 0],
      // 11: bright yellow
      [0, 0, 255],
      // 12: bright blue
      [255, 0, 255],
      // 13: bright magenta
      [0, 255, 255],
      // 14: bright cyan
      [255, 255, 255]
      // 15: bright white
    ];
    return standard[index];
  }
  if (index < 232) {
    const i = index - 16;
    const r = Math.floor(i / 36);
    const g = Math.floor(i % 36 / 6);
    const b = i % 6;
    return [
      r === 0 ? 0 : 55 + r * 40,
      g === 0 ? 0 : 55 + g * 40,
      b === 0 ? 0 : 55 + b * 40
    ];
  }
  const gray = (index - 232) * 10 + 8;
  return [gray, gray, gray];
}
function extractRgbFromAnsi(ansiCode) {
  const rgbMatch = ansiCode.match(/\x1b\[38;2;(\d+);(\d+);(\d+)m/);
  if (rgbMatch) {
    return [parseInt(rgbMatch[1], 10), parseInt(rgbMatch[2], 10), parseInt(rgbMatch[3], 10)];
  }
  const ansi256Match = ansiCode.match(/\x1b\[38;5;(\d+)m/);
  if (ansi256Match) {
    return ansi256ToRgb(parseInt(ansi256Match[1], 10));
  }
  const basicMatch = ansiCode.match(/\x1b\[(\d+)m/);
  if (basicMatch) {
    const code = parseInt(basicMatch[1], 10);
    if (code >= 30 && code <= 37) {
      return ansi256ToRgb(code - 30);
    }
    if (code >= 90 && code <= 97) {
      return ansi256ToRgb(code - 90 + 8);
    }
  }
  return null;
}
function parseColorToRgb(color, themeColors) {
  if (color.startsWith("#")) {
    const hex = color.slice(1);
    if (hex.length === 3) {
      return [
        parseInt(hex[0] + hex[0], 16),
        parseInt(hex[1] + hex[1], 16),
        parseInt(hex[2] + hex[2], 16)
      ];
    }
    return [
      parseInt(hex.slice(0, 2), 16),
      parseInt(hex.slice(2, 4), 16),
      parseInt(hex.slice(4, 6), 16)
    ];
  }
  if (color.startsWith("\x1B[")) {
    const rgb = extractRgbFromAnsi(color);
    if (rgb) return rgb;
  }
  if (themeColors) {
    const themeColor = themeColors[color.toLowerCase()];
    if (typeof themeColor === "string") {
      const rgb = extractRgbFromAnsi(themeColor);
      if (rgb) return rgb;
    }
  }
  const named = NAMED_COLORS[color.toLowerCase()];
  if (named) return named;
  return [255, 255, 255];
}
function interpolateRgb(color1, color2, t) {
  return [
    Math.round(color1[0] + (color2[0] - color1[0]) * t),
    Math.round(color1[1] + (color2[1] - color1[1]) * t),
    Math.round(color1[2] + (color2[2] - color1[2]) * t)
  ];
}
function getColorAtPosition(gradient, position, themeColors) {
  const { colors } = gradient;
  if (colors.length === 0) return [255, 255, 255];
  if (colors.length === 1) return parseColorToRgb(colors[0], themeColors);
  const clampedPosition = Math.max(0, Math.min(1, position));
  const scaledPosition = clampedPosition * (colors.length - 1);
  const lowerIndex = Math.floor(scaledPosition);
  const upperIndex = Math.min(lowerIndex + 1, colors.length - 1);
  const localT = scaledPosition - lowerIndex;
  const color1 = parseColorToRgb(colors[lowerIndex], themeColors);
  const color2 = parseColorToRgb(colors[upperIndex], themeColors);
  return interpolateRgb(color1, color2, localT);
}
function applyGradientToText(text, gradient, themeColors) {
  const { colors } = gradient;
  if (colors.length === 0 || text.length === 0) return text;
  const chars = [...text];
  if (colors.length === 1) {
    const [r, g, b] = parseColorToRgb(colors[0], themeColors);
    return `\x1B[38;2;${r};${g};${b}m${text}\x1B[0m`;
  }
  let result = "";
  for (let i = 0; i < chars.length; i++) {
    let position = chars.length > 1 ? i / (chars.length - 1) : 0;
    if (gradient.reverse) position = 1 - position;
    const [r, g, b] = getColorAtPosition(gradient, position, themeColors);
    result += `\x1B[38;2;${r};${g};${b}m${chars[i]}\x1B[39m`;
  }
  return result;
}
function interpolateGradientColor(gradient, index, total, themeColors) {
  const { colors } = gradient;
  if (colors.length === 0) return "";
  if (colors.length === 1) return colors[0];
  const position = total > 1 ? index / (total - 1) : 0;
  const [r, g, b] = getColorAtPosition(gradient, position, themeColors);
  return `#${r.toString(16).padStart(2, "0")}${g.toString(16).padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
}

// formatters/formatter.ts
var ChartFormatter = class {
  constructor(options) {
    this.options = options;
    this.colors = style({ theme: options.theme });
  }
  calculateBarDimensions(barCount, totalSize, charWidth = 1) {
    const alignment = this.options.alignBars || "justify";
    if (alignment !== "justify") {
      return {
        barSize: this.options.barSize || 1,
        padding: this.options.padding || 0
      };
    }
    const hasExplicitPadding = this.options.padding !== void 0 && this.options.padding > 0;
    if (hasExplicitPadding && barCount > 1) {
      if (this.options.barSize !== void 0) {
        const minSize = this.options.barSize * charWidth * barCount + this.options.padding * (barCount - 1);
        const extraPadding = Math.max(0, totalSize - minSize);
        const paddingPerGap = Math.floor(extraPadding / (barCount - 1));
        return {
          barSize: this.options.barSize,
          padding: this.options.padding + paddingPerGap
        };
      }
      const totalPaddingSize = this.options.padding * (barCount - 1);
      const availableForBars = totalSize - totalPaddingSize;
      return {
        barSize: Math.max(1, Math.floor(availableForBars / (barCount * charWidth))),
        padding: this.options.padding
      };
    }
    return this.calculateDefaultDimensions(barCount, totalSize, charWidth);
  }
  calculateAlignmentPadding(barsWidth, totalSize, alignment) {
    if (alignment === "justify") {
      return { leadingPad: 0, trailingPad: 0 };
    }
    const extraSpace = Math.max(0, totalSize - barsWidth);
    if (alignment === "right" || alignment === "bottom") {
      return { leadingPad: extraSpace, trailingPad: 0 };
    }
    if (alignment === "center") {
      const leading = Math.floor(extraSpace / 2);
      return { leadingPad: leading, trailingPad: extraSpace - leading };
    }
    return { leadingPad: 0, trailingPad: extraSpace };
  }
  formatValueWithDecimals(value) {
    const formattedValue = this.options.valueLabelsFloatingPoint !== void 0 ? value.toFixed(this.options.valueLabelsFloatingPoint) : String(value);
    return this.options.valueLabelFormat ? this.options.valueLabelFormat([formattedValue]) : formattedValue;
  }
  formatStackedValueLabel(segments) {
    const values = segments.map((s) => {
      return this.options.valueLabelsFloatingPoint !== void 0 ? s.value.toFixed(this.options.valueLabelsFloatingPoint) : String(s.value);
    });
    return this.options.valueLabelFormat ? this.options.valueLabelFormat(values) : values.join("|");
  }
  formatPercentage(point) {
    return this.options.percentage ? `(${point.percentage.toFixed(2)}%)` : "";
  }
  colorify(txt, color) {
    if (!color) return txt;
    const normalizedColor = typeof color === "string" ? normalizeColor(color) : color;
    if (isGradientObject(normalizedColor)) {
      return this.applyGradient(txt, normalizedColor);
    }
    if (Array.isArray(normalizedColor)) {
      return this.colors.rgb(...normalizedColor)`${txt}`;
    }
    if (typeof normalizedColor === "string") {
      if (normalizedColor.includes("#")) {
        return this.colors.hex(normalizedColor)`${txt}`;
      }
      if (normalizedColor.match(/[0-9]/)) {
        return this.colors.ansi(normalizedColor)`${txt}`;
      }
      if (!this.options.richLabels) {
        const ansiCode = this.colors.colors[normalizedColor];
        if (ansiCode) return `${ansiCode}${txt}${this.colors.colors.reset}`;
      }
      return this.colors[normalizedColor]`${txt}`;
    }
    return txt;
  }
  isGradient(color) {
    const normalized = typeof color === "string" ? normalizeColor(color) : color;
    return isGradientObject(normalized);
  }
  normalizeGradient(color) {
    if (typeof color === "string") {
      return normalizeColor(color);
    }
    return color;
  }
  getThemeColors() {
    return this.colors.colors;
  }
  getColorAtPosition(gradient, position) {
    return getColorAtPosition(gradient, position, this.getThemeColors());
  }
  getLabelColorForBar(color, barIndex, totalBars) {
    if (!color) return void 0;
    const normalizedColor = typeof color === "string" ? normalizeColor(color) : color;
    if (!isGradientObject(normalizedColor)) return normalizedColor;
    const { direction = "horizontal", reverse = false } = normalizedColor;
    const chartOrientation = this.options.orientation || "horizontal";
    let position;
    if (chartOrientation === "horizontal") {
      if (direction === "horizontal") {
        position = 0;
      } else if (direction === "diagonal") {
        const vPos = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
        position = vPos / 2;
      } else {
        position = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
      }
    } else {
      if (direction === "vertical") {
        position = 1;
      } else if (direction === "diagonal") {
        const hPos = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
        position = (hPos + 1) / 2;
      } else {
        position = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
      }
    }
    if (reverse) {
      position = 1 - position;
    }
    const [r, g, b] = this.getColorAtPosition(normalizedColor, position);
    return `#${r.toString(16).padStart(2, "0")}${g.toString(16).padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
  }
  getValueLabelColorForBar(color, barIndex, totalBars, barEndPosition) {
    if (!color) return void 0;
    const normalizedColor = typeof color === "string" ? normalizeColor(color) : color;
    if (!isGradientObject(normalizedColor)) return normalizedColor;
    const { direction = "horizontal", reverse = false } = normalizedColor;
    const chartOrientation = this.options.orientation || "horizontal";
    let position;
    if (chartOrientation === "horizontal") {
      if (direction === "horizontal") {
        position = barEndPosition !== void 0 ? barEndPosition : 1;
      } else if (direction === "diagonal") {
        const hPos = barEndPosition !== void 0 ? barEndPosition : 1;
        const vPos = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
        position = (hPos + vPos) / 2;
      } else {
        position = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
      }
    } else {
      if (direction === "vertical") {
        position = barEndPosition !== void 0 ? barEndPosition : 0;
      } else if (direction === "diagonal") {
        const hPos = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
        const vPos = barEndPosition !== void 0 ? barEndPosition : 0;
        position = (hPos + vPos) / 2;
      } else {
        position = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
      }
    }
    if (reverse) {
      position = 1 - position;
    }
    const [r, g, b] = this.getColorAtPosition(normalizedColor, position);
    return `#${r.toString(16).padStart(2, "0")}${g.toString(16).padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
  }
  applyGradientWithContext(text, gradient, charIndex, totalChars, barIndex, totalBars, rowIndex = 0, totalRows = 1) {
    const { colors, direction = "horizontal", reverse = false } = gradient;
    if (colors.length === 0 || text.length === 0) return text;
    const chars = [...text];
    let result = "";
    for (let i = 0; i < chars.length; i++) {
      let position;
      if (direction === "vertical") {
        position = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
      } else if (direction === "diagonal") {
        const globalCharIndex = charIndex + i;
        const hPos = totalChars > 1 ? globalCharIndex / (totalChars - 1) : 0;
        const vPos = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
        position = (hPos + vPos) / 2;
      } else {
        const globalCharIndex = charIndex + i;
        position = totalChars > 1 ? globalCharIndex / (totalChars - 1) : 0;
      }
      if (reverse) {
        position = 1 - position;
      }
      const [r, g, b] = this.getColorAtPosition(gradient, position);
      result += `\x1B[38;2;${r};${g};${b}m${chars[i]}\x1B[39m`;
    }
    return result;
  }
  applyGradient(text, gradient) {
    return applyGradientToText(text, gradient, this.getThemeColors());
  }
  applyDecorators(text) {
    if (!this.options.richLabels) return text;
    const decorators = {
      "*": { symbol: 1, reset: 22 },
      // bold
      "~": { symbol: 2, reset: 22 },
      // dim
      "%": { symbol: 3, reset: 23 },
      // italic
      "!": { symbol: 4, reset: 24 },
      // underline
      "^": { symbol: 5, reset: 25 },
      // blink
      "@": { symbol: 7, reset: 27 },
      // invert
      "#": { symbol: 8, reset: 28 },
      // hidden
      "$": { symbol: 9, reset: 29 }
      // strikeout
    };
    let result = text;
    for (const [char, { symbol, reset }] of Object.entries(decorators)) {
      const escaped = char.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      const regex = new RegExp(`${escaped}(.+?)${escaped}`, "gs");
      result = result.replace(regex, `\x1B[${symbol}m$1\x1B[${reset}m`);
    }
    return result;
  }
  stripStyle(label) {
    return label.replace(/\x1b\[[0-9;]*m/g, "");
  }
  getTitleConfig() {
    const { title } = this.options;
    if (!title) return void 0;
    if (typeof title === "string") {
      return { text: title };
    }
    return title;
  }
  parseTitlePadding(padding) {
    if (padding === void 0) {
      return { top: 0, right: 0, bottom: 0, left: 0 };
    }
    if (typeof padding === "number") {
      return { top: padding, right: padding, bottom: padding, left: padding };
    }
    if (padding.length === 2) {
      const [vertical, horizontal] = padding;
      return { top: vertical, right: horizontal, bottom: vertical, left: horizontal };
    }
    const [top, right, bottom, left] = padding;
    return { top, right, bottom, left };
  }
  formatTitle(width) {
    const titleConfig = this.getTitleConfig();
    if (!titleConfig) return "";
    const { text, align = "left", color, padding: titlePadding } = titleConfig;
    const { top, right, bottom, left } = this.parseTitlePadding(titlePadding);
    let formattedText = text;
    if (color === "gradient") {
      const chartColor = this.isGradient(this.options.color) ? this.options.color : this.isGradient(this.options._seriesColors?.[0]) ? this.options._seriesColors[0] : this.isGradient(this.options._bullColor) ? this.options._bullColor : void 0;
      if (chartColor) {
        formattedText = this.colorify(text, chartColor);
      }
    } else if (color) {
      formattedText = this.colorify(text, color);
    }
    if (align === "left" && !titlePadding) {
    } else if (align !== "left") {
      if (width && width > text.length) {
        const availableSpace = width - text.length;
        switch (align) {
          case "center": {
            const leftPad = Math.floor(availableSpace / 2);
            const rightPad = availableSpace - leftPad;
            formattedText = " ".repeat(leftPad) + formattedText + " ".repeat(rightPad);
            break;
          }
          case "right":
            formattedText = " ".repeat(availableSpace) + formattedText;
            break;
        }
      }
    } else if (titlePadding) {
      formattedText = " ".repeat(left) + formattedText + " ".repeat(right);
    }
    const lines = [];
    for (let i = 0; i < top; i++) {
      lines.push("");
    }
    lines.push(formattedText);
    for (let i = 0; i < bottom; i++) {
      lines.push("");
    }
    return lines.join("\n");
  }
};
var formatter_default = ChartFormatter;

// formatters/horizontal.ts
var HorizontalChartFormatter = class extends formatter_default {
  constructor(options) {
    super(options);
  }
  calculateDefaultDimensions(barCount, totalSize, _charWidth) {
    const hasPadding = this.options.padding !== void 0;
    const chartPadding = hasPadding ? this.options.padding : 0;
    const defaultPadding = Math.floor((totalSize - chartPadding) / barCount);
    const barSize = this.options.barSize || defaultPadding || 1;
    const padding = hasPadding ? chartPadding : defaultPadding;
    return { barSize, padding };
  }
  format(chart) {
    const barCount = chart.size;
    const { barSize, padding } = this.calculateBarDimensions(barCount, this.options.height);
    const labels = [];
    const chartArray = Array.from(chart.values());
    const lines = Array.from(chart.entries()).map(([i, point]) => {
      const isLast = Number(i) === chart.size - 1;
      const barIndex = Number(i);
      labels.push(this.formatLabel(point, this.options.structure.y, barIndex, barCount));
      const ctx = {
        chart: chartArray,
        point,
        index: barIndex,
        barSize,
        padding: isLast ? 0 : padding
      };
      return this.formatLineWithContext(ctx);
    });
    const alignment = this.options.alignBars || "justify";
    const { leadingPad, trailingPad } = this.calculateVerticalPadding(barCount, barSize, padding, alignment);
    const topPadLines = this.buildAlignmentPadLines(leadingPad, labels);
    const bottomPadLines = this.buildAlignmentPadLines(trailingPad, labels);
    const strippedLabels = labels.map((l) => this.stripStyle(l));
    const maxLabelWidth = Math.max(...strippedLabels.map((label) => label.length - 1));
    const totalChartWidth = maxLabelWidth + 1 + this.options.width;
    return [
      this.formatChartLabel(totalChartWidth),
      ...topPadLines,
      ...lines,
      ...bottomPadLines,
      this.formatBottom(labels)
    ].join("\n");
  }
  calculateVerticalPadding(barCount, barSize, padding, alignment) {
    if (alignment === "justify" || alignment === "left" || alignment === "right") {
      return { leadingPad: 0, trailingPad: 0 };
    }
    const barsHeight = barCount * (barSize + padding) - padding;
    return this.calculateAlignmentPadding(barsHeight, this.options.height, alignment);
  }
  buildAlignmentPadLines(count, labels) {
    if (count <= 0) return [];
    const strippedLabels = labels.map((l) => this.stripStyle(l));
    const maxLabelLength = Math.max(...strippedLabels.map((label) => label.length - 1));
    const axisChar = this.formatStructure(this.options.structure.axis);
    const linePad = this.options.labels ? this.pad(maxLabelLength) : "";
    return Array.from({ length: count }, () => linePad + axisChar);
  }
  formatLineWithContext(ctx) {
    const label = this.formatLabel(ctx.point, this.options.structure.y, ctx.index, ctx.chart.length);
    const bar = this.formatBarWithContext(ctx, label);
    return `${label}${bar}`;
  }
  formatBarWithContext(ctx, label) {
    if (ctx.point.segments?.length) {
      return this.formatStackedBar(ctx.point, label, ctx.barSize, ctx.padding, ctx.index, ctx.chart.length, ctx.chart);
    }
    return this.formatSingleBarWithContext(ctx, label);
  }
  formatSingleBarWithContext(ctx, label) {
    const repeat = ctx.point.scaled / this.options.char.length;
    const color = ctx.point.color || this.options.color || "";
    const barChars = this.options.char?.repeat(repeat);
    const totalBars = ctx.chart.length;
    const currentMaxBarLength = Math.max(...ctx.chart.map((p) => Math.floor(p.scaled / this.options.char.length)));
    const finalMaxBarLength = this.options._finalMaxBarLength ?? currentMaxBarLength;
    const barEndPosition = finalMaxBarLength > 1 ? (repeat - 1) / (finalMaxBarLength - 1) : 0;
    if (this.isGradient(color)) {
      const gradient = this.normalizeGradient(color);
      const direction = gradient.direction || "horizontal";
      const reverse = gradient.reverse || false;
      const fill2 = this.options.fillColor === "auto" ? this.formatGradientFill(ctx.point, gradient, ctx.index, totalBars, finalMaxBarLength, repeat) : this.formatFill(ctx.point);
      if (direction === "vertical" || direction === "diagonal") {
        const totalChars = finalMaxBarLength * totalBars;
        const chars = [...barChars];
        let result = "";
        for (let i = 0; i < chars.length; i++) {
          const globalCharIndex = ctx.index * finalMaxBarLength + i;
          let position;
          if (direction === "diagonal") {
            const hPos = finalMaxBarLength > 1 ? i / (finalMaxBarLength - 1) : 0;
            const vPos = totalChars > 1 ? globalCharIndex / (totalChars - 1) : 0;
            position = (hPos + vPos) / 2;
          } else {
            position = totalChars > 1 ? globalCharIndex / (totalChars - 1) : 0;
          }
          if (reverse) position = 1 - position;
          const [r, g, b] = this.getColorAtPosition(gradient, position);
          result += `\x1B[38;2;${r};${g};${b}m${chars[i]}\x1B[39m`;
        }
        const barContent3 = result + fill2;
        return this.scaleBar(barContent3, ctx.point, label, ctx.barSize, ctx.padding, ctx.index, totalBars, void 0, barEndPosition);
      }
      const coloredBar = this.applyGradientWithContext(barChars, gradient, 0, finalMaxBarLength, ctx.index, totalBars);
      const barContent2 = coloredBar + fill2;
      return this.scaleBar(barContent2, ctx.point, label, ctx.barSize, ctx.padding, ctx.index, totalBars, void 0, barEndPosition);
    }
    const fill = this.formatFill(ctx.point);
    if (this.options.fillColor) {
      const coloredBar = color ? this.colorify(barChars, color) : barChars;
      const barContent2 = coloredBar + fill;
      return this.scaleBar(barContent2, ctx.point, label, ctx.barSize, ctx.padding, ctx.index, totalBars, void 0, barEndPosition);
    }
    const barContent = barChars + fill;
    return this.scaleBar(barContent, ctx.point, label, ctx.barSize, ctx.padding, ctx.index, totalBars, color, barEndPosition);
  }
  formatStackedBar(point, label, barSize, padding, barIndex, totalBars, chart) {
    const combinedBar = this.buildStackedBarContent(point.segments);
    const barWithFill = combinedBar + this.formatFill(point);
    const valueLabel = this.getStackedValueLabel(point);
    const maxBarLength = Math.max(...chart.map((p) => Math.floor(p.scaled / this.options.char.length)));
    const barLength = Math.floor(point.scaled / this.options.char.length);
    const barEndPosition = maxBarLength > 1 ? (barLength - 1) / (maxBarLength - 1) : 0;
    return this.buildScaledBar(barWithFill, point, label, barSize, padding, valueLabel, barIndex, totalBars, void 0, barEndPosition);
  }
  buildStackedBarContent(segments) {
    return segments.map((seg) => this.formatSegment(seg)).join("");
  }
  formatSegment(segment) {
    const repeat = Math.floor(segment.scaled / this.options.char.length);
    const segmentBar = this.options.char?.repeat(repeat);
    return segment.color ? this.colorify(segmentBar, segment.color) : segmentBar;
  }
  getStackedValueLabel(point) {
    if ((this.options.valueLabels || this.options.stackValueLabels) && point.segments) {
      return this.formatStackedValueLabel(point.segments);
    }
    return this.formatValueWithDecimals(point.value);
  }
  scaleBar(bar, point, label, barSize, padding, barIndex, totalBars, color, barEndPosition) {
    const valueLabel = this.formatValueWithDecimals(point.value);
    return this.buildScaledBar(bar, point, label, barSize, padding, valueLabel, barIndex, totalBars, color, barEndPosition);
  }
  buildScaledBar(bar, point, label, barSize, padding, valueLabel, barIndex, totalBars, color, barEndPosition) {
    const space = this.calculateLabelSpace(label);
    const barLines = this.buildBarLines(bar, point, space, barSize, valueLabel, barIndex, totalBars, color, barEndPosition);
    const paddingLines = this.buildPaddingLines(space, padding);
    return [...barLines, ...paddingLines].join("\n");
  }
  calculateLabelSpace(label) {
    const strippedLabel = this.stripStyle(label);
    const naked = this.options.naked ? 0 : 1;
    return strippedLabel.length - naked;
  }
  buildBarLines(bar, point, space, barSize, valueLabel, barIndex, totalBars, color, barEndPosition) {
    return Array.from(
      { length: barSize },
      (_, index) => this.buildBarLine(bar, point, space, index, valueLabel, barIndex, totalBars, color, barEndPosition)
    );
  }
  buildBarLine(bar, point, space, index, valueLabel, barIndex, totalBars, color, barEndPosition) {
    const axisChar = this.formatStructure(this.options.structure.axis);
    const linePad = index !== 0 ? this.pad(space) + axisChar : "";
    const shouldShowValue = (this.options.valueLabels || this.options.stackValueLabels) && index === 0;
    const barContent = color ? this.colorify(bar, color) : bar;
    let lineContent;
    if (shouldShowValue) {
      let valueLabelColor;
      if (this.options.fill && this.options.fillColor) {
        if (this.options.fillColor === "auto") {
          const pointColor = point.color || this.options.color;
          valueLabelColor = this.getValueLabelColorForBar(pointColor, barIndex, totalBars, 1);
        } else {
          valueLabelColor = this.options.fillColor;
        }
      } else {
        const pointColor = point.color || this.options.color;
        valueLabelColor = this.getValueLabelColorForBar(pointColor, barIndex, totalBars, barEndPosition);
      }
      const coloredValueLabel = valueLabelColor ? this.colorify(valueLabel, valueLabelColor) : valueLabel;
      lineContent = barContent + this.pad(1) + coloredValueLabel;
    } else {
      lineContent = barContent;
    }
    return linePad + lineContent;
  }
  buildPaddingLines(space, padding) {
    const axisChar = this.formatStructure(this.options.structure.axis);
    const linePad = this.options.labels ? this.pad(space) : "";
    return Array.from({ length: padding }, () => linePad + axisChar);
  }
  formatFill(point, lastBarColor) {
    if (!this.options.fill) return "";
    const diff = this.options.width - point.scaled;
    const fillLength = this.options.fill.length;
    let fillStr = "";
    if (this.options.scale) {
      const width = Math.floor(this.options.width - Math.floor(point.scaled));
      const fillCount = Math.floor(width / fillLength);
      fillStr = fillCount > 0 ? this.options.fill.repeat(fillCount) : "";
    } else {
      const fillCount = Math.floor(diff / fillLength);
      fillStr = fillCount > 0 ? this.options.fill.repeat(fillCount) : "";
    }
    if (!fillStr) return "";
    if (this.options.fillColor) {
      if (this.options.fillColor === "auto") {
        if (lastBarColor) {
          return `\x1B[38;2;${lastBarColor}m${fillStr}\x1B[39m`;
        }
        const barColor = point.color || this.options.color;
        if (barColor && !this.isGradient(barColor)) {
          return this.colorify(fillStr, barColor);
        }
      } else {
        return this.colorify(fillStr, this.options.fillColor);
      }
    }
    return fillStr;
  }
  formatGradientFill(point, gradient, barIndex, totalBars, finalMaxBarLength, currentBarLength) {
    if (!this.options.fill) return "";
    const diff = this.options.width - point.scaled;
    const fillLength = this.options.fill.length;
    let fillCount = 0;
    if (this.options.scale) {
      const width = Math.floor(this.options.width - Math.floor(point.scaled));
      fillCount = Math.floor(width / fillLength);
    } else {
      fillCount = Math.floor(diff / fillLength);
    }
    if (fillCount <= 0) return "";
    const direction = gradient.direction || "horizontal";
    const reverse = gradient.reverse || false;
    let result = "";
    for (let i = 0; i < fillCount; i++) {
      const charIndexInBar = Math.floor(currentBarLength) + i;
      let position;
      if (direction === "vertical") {
        const totalChars = finalMaxBarLength * totalBars;
        const globalCharIndex = barIndex * finalMaxBarLength + charIndexInBar;
        position = totalChars > 1 ? globalCharIndex / (totalChars - 1) : 0;
      } else if (direction === "diagonal") {
        const hPos = finalMaxBarLength > 1 ? charIndexInBar / (finalMaxBarLength - 1) : 0;
        const vPos = totalBars > 1 ? barIndex / (totalBars - 1) : 0;
        position = (hPos + vPos) / 2;
      } else {
        position = finalMaxBarLength > 1 ? charIndexInBar / (finalMaxBarLength - 1) : 0;
      }
      if (reverse) position = 1 - position;
      const [r, g, b] = this.getColorAtPosition(gradient, position);
      result += `\x1B[38;2;${r};${g};${b}m${this.options.fill}\x1B[39m`;
    }
    return result;
  }
  formatLabel(point, key, barIndex, totalBars) {
    const percentage = this.formatPercentage(point);
    const baseLabel = percentage ? `${point.label} ${percentage}` : point.label;
    const formatted = this.options.labelFormat ? this.options.labelFormat(baseLabel) : baseLabel;
    const label = this.applyDecorators(formatted);
    const space = this.formatLabelSpace(label);
    if (!this.options.labels) {
      return this.formatStructure(this.options.structure.axis);
    }
    const pointColor = point.color || this.options.color;
    const labelColor = this.getLabelColorForBar(pointColor, barIndex, totalBars);
    const coloredLabel = this.options.colorLabels && labelColor ? this.colorify(label, labelColor) : label;
    return `${coloredLabel}${space}${this.formatStructure(key)}`;
  }
  formatLabelSpace(label) {
    if (!this.options.max.label) return "";
    const addOne = this.offsetPercentage();
    const space = this.options.max.label - this.stripStyle(label).length + addOne;
    return this.pad(space);
  }
  formatChartLabel(totalWidth) {
    return this.formatTitle(totalWidth);
  }
  formatStructure(structChar) {
    if (this.options.naked) return "";
    return this.colors.colors.reset + structChar;
  }
  formatBottom(labels) {
    if (this.options.naked) return "";
    const strippedLabels = labels.map((l) => this.stripStyle(l));
    const max = Math.max(...strippedLabels.map((label) => label.length - 1));
    return this.pad(max) + this.options.structure.bottomLeft + this.options.structure.x.repeat(this.options.width);
  }
  pad(space) {
    return " ".repeat(space);
  }
  offsetPercentage() {
    return this.options.labels && !this.options.percentage ? 1 : 0;
  }
};
var horizontal_default = HorizontalChartFormatter;

// validator/validator.ts
var ChartValidator = class {
  constructor(options) {
    this.options = options;
  }
  error(text) {
    return new Error(text);
  }
  validate(data) {
    if (!Array.isArray(data)) throw new Error("Input data must be an array");
    if (typeof data[0] === "string") throw new Error("Input values must be numbers. e.g [1, 2, 3] or [{value: 1}, {value: 2}]");
    if (!data.length) throw new Error("No data provided");
    if (this.options.barSize === 0) throw new Error("barSize cannot be 0");
    this.validateStackedData(data);
  }
  isStackedPoint(point) {
    return typeof point !== "number" && Array.isArray(point.value);
  }
  validateStackedData(data) {
    if (this.options.type === "candlestick" || this.options.type === "status") return;
    const stackedPoints = data.filter((d) => this.isStackedPoint(d));
    if (stackedPoints.length === 0) return;
    const segmentCounts = stackedPoints.map((d) => {
      const value = d.value;
      return value.length;
    });
    const uniqueCounts = new Set(segmentCounts);
    if (uniqueCounts.size > 1) {
      throw new Error("All stacked bars must have the same number of segments");
    }
    const segmentCount = segmentCounts[0];
    if (this.options.stackColors && this.options.stackColors.length !== segmentCount) {
      console.warn(`stackColors count (${this.options.stackColors.length}) does not match segment count (${segmentCount})`);
    }
  }
};
var validator_default = ChartValidator;

// processor/processor.ts
var AUTO_COLORS = [
  "red",
  "green",
  "yellow",
  "blue",
  "purple",
  "cyan",
  "pink",
  "orange",
  "marine"
];
var ChartProcessor = class {
  constructor(options) {
    this.validator = new validator_default(options);
    if (options.color && !Array.isArray(options.color)) {
      options.color = normalizeColor(options.color);
    }
    this.options = options;
  }
  isStackedPoint(point) {
    if (this.isCandlestick()) return false;
    if (this.isStatus()) return false;
    return typeof point !== "number" && Array.isArray(point.value);
  }
  isCandlestick() {
    return this.options.type === "candlestick";
  }
  isStatus() {
    return this.options.type === "status";
  }
  getStatus(point) {
    if (typeof point === "number") return String(point);
    const v = point.value;
    if (typeof v === "number") return String(v);
    if (typeof v === "string") return v;
    return void 0;
  }
  /**
   * Status row mode: when `value` is an array of numbers/strings, return one
   * status key per cell. Used by the status formatter to render a row of
   * colored cells with the point's `label` on the left.
   */
  getStatuses(point) {
    if (typeof point === "number") return void 0;
    const v = point.value;
    if (!Array.isArray(v)) return void 0;
    return v.map((entry) => {
      if (typeof entry === "number" || typeof entry === "string") return String(entry);
      return String(entry.value);
    });
  }
  getOHLC(point) {
    if (typeof point === "number") return void 0;
    const v = point.value;
    if (!Array.isArray(v) || v.length < 4) return void 0;
    const num = (x) => typeof x === "number" ? x : x.value;
    return [num(v[0]), num(v[1]), num(v[2]), num(v[3])];
  }
  isGradient(color) {
    const normalized = typeof color === "string" ? normalizeColor(color) : color;
    return isGradientObject(normalized);
  }
  interpolateGradient(gradient, index, total) {
    const result = interpolateGradientColor(gradient, index, total);
    return result || this.options.color || "";
  }
  getSegmentValue(segment) {
    return typeof segment === "number" ? segment : segment.value;
  }
  getPointValue(point) {
    if (this.isStatus()) return 0;
    if (this.isCandlestick()) {
      const ohlc = this.getOHLC(point);
      if (ohlc) return ohlc[3];
    }
    if (typeof point === "number") return point;
    if (this.isStackedPoint(point)) {
      return point.value.reduce(
        (sum, seg) => sum + this.getSegmentValue(seg),
        0
      );
    }
    return point.value;
  }
  getPointLabel(point, value) {
    if (typeof point === "number") {
      return this.isStatus() ? "" : point.toString();
    }
    if (point.label) return point.label;
    if (this.isCandlestick()) return value.toString();
    if (this.isStatus()) return "";
    if (this.isStackedPoint(point)) return value.toString();
    return point.value.toString();
  }
  getPointColor(point) {
    if (typeof point === "number") {
      const color2 = this.options.color;
      if (typeof color2 === "string") return color2 || void 0;
      return void 0;
    }
    if (Array.isArray(point.color)) {
      const optColor = this.options.color;
      return point.color[0] || (typeof optColor === "string" ? optColor : void 0) || void 0;
    }
    if (typeof point.color === "string") return point.color || void 0;
    const color = this.options.color;
    return (typeof color === "string" ? color : void 0) || void 0;
  }
  getPointSegments(point, value) {
    if (!this.isStackedPoint(point)) return void 0;
    const pointColors = this.getPointColorArray(point);
    return this.processSegments(point.value, value, pointColors);
  }
  getPointColorArray(point) {
    if (typeof point === "number") return void 0;
    if (Array.isArray(point.color)) return point.color;
    return void 0;
  }
  calculateTotal(data) {
    return data.reduce((sum, point) => sum + this.getPointValue(point), 0);
  }
  calculateData(data) {
    const total = this.calculateTotal(data);
    if (this.isStatus()) {
      return data.reduce((sum, point) => sum + this.getPointValue(point), 0);
    }
    data.forEach((point) => {
      if (this.isCandlestick()) {
        const ohlc = this.getOHLC(point);
        if (ohlc) {
          const [, high, low] = ohlc;
          this.options.max.value = Math.max(high, this.options.max.value);
          this.options.max.min = this.options.max.min === void 0 ? low : Math.min(low, this.options.max.min);
          return;
        }
      }
      const value = this.getPointValue(point);
      this.options.max.value = Math.max(value, this.options.max.value);
      if (this.options.max.min === void 0) {
        this.options.max.min = value;
      } else {
        this.options.max.min = Math.min(value, this.options.max.min);
      }
    });
    return data.reduce((sum, point) => {
      const value = this.getPointValue(point);
      const label = this.getPointLabel(point, value);
      const percentage = this.percentage(value, total);
      const percentageLength = this.options.percentage ? (isNaN(percentage) ? 3 : percentage.toFixed(2).length) + 5 : 0;
      const decoratorSymbols = /([*~%!^@#])(.+?)\1/g;
      const visualLabel = this.options.richLabels !== false ? label.replace(decoratorSymbols, "$2") : label;
      let baseLabelLength = this.options.percentage ? visualLabel.length + percentageLength : visualLabel.length;
      if (this.options.labelFormat) {
        const fullLabel = this.options.percentage ? `${label} (${percentage.toFixed(2)}%)` : label;
        const formattedLabel = this.options.labelFormat(fullLabel);
        const strippedFormatted = formattedLabel.replace(/\x1b\[[0-9;]*m/g, "");
        baseLabelLength = strippedFormatted.length;
      }
      if (this.options.labels) {
        if (this.options._maxLabel !== void 0) {
          this.options.max.label = this.options._maxLabel;
        } else {
          this.options.max.label = Math.max(baseLabelLength, this.options.max.label);
        }
      }
      this.options.max.scaled = Math.max(this.scale(value), this.options.max.scaled);
      return sum + value;
    }, 0);
  }
  percentage(value, total) {
    if (!this.options.percentage) return 0;
    return value / total * 100;
  }
  scale(value) {
    const size = this.options.orientation === "vertical" ? this.options.height : this.options.width;
    const { scale, max } = this.options;
    if (scale === "auto" || scale === void 0) {
      return Math.ceil(value / max.value * size);
    }
    if (scale === "relative" && max.min !== void 0 && max.min !== max.value) {
      const range = max.value - max.min;
      const normalized = value - max.min;
      return Math.ceil((normalized + 1) / (range + 1) * size);
    }
    if (scale === "relative-zero" && max.min !== void 0 && max.min !== max.value) {
      const range = max.value - max.min;
      const normalized = value - max.min;
      return Math.ceil(normalized / range * size);
    }
    if (typeof scale === "number" && scale > 0) {
      return Math.min(Math.round(value / scale), size);
    }
    return Math.min(value, size);
  }
  processSegments(segments, totalBarValue, pointColors) {
    const totalScaled = this.scale(totalBarValue);
    const rawSegments = segments.map((seg, index) => {
      const value = this.getSegmentValue(seg);
      const color = this.getSegmentColor(seg, index, pointColors);
      const proportion = totalBarValue > 0 ? value / totalBarValue : 0;
      const scaled = Math.floor(proportion * totalScaled);
      const percentage = totalBarValue > 0 ? value / totalBarValue * 100 : 0;
      return { value, color, scaled, percentage };
    });
    const sumScaled = rawSegments.reduce((sum, seg) => sum + seg.scaled, 0);
    const remainder = totalScaled - sumScaled;
    if (remainder > 0) {
      const largestIndex = rawSegments.reduce((maxIdx, seg, idx, arr) => {
        return seg.value > arr[maxIdx].value ? idx : maxIdx;
      }, 0);
      rawSegments[largestIndex].scaled += remainder;
    }
    return rawSegments;
  }
  applyAutoColor(data) {
    return data.map((point, index) => {
      const color = AUTO_COLORS[index % AUTO_COLORS.length];
      if (typeof point === "number") {
        return { value: point, color };
      }
      if (point.color) {
        return point;
      }
      return { ...point, color };
    });
  }
  applyAutoStackColors(data) {
    if (!this.options.stackColors || this.options.stackColors.length === 0) {
      let maxSegments = 0;
      data.forEach((point) => {
        if (this.isStackedPoint(point)) {
          maxSegments = Math.max(maxSegments, point.value.length);
        }
      });
      if (maxSegments > 0) {
        this.options.stackColors = AUTO_COLORS.slice(0, maxSegments);
      }
    }
  }
  applyGradient(data, gradient) {
    return data.map((point, index) => {
      const color = this.interpolateGradient(gradient, index, data.length);
      if (typeof point === "number") {
        return { value: point, color };
      }
      if (point.color) {
        return point;
      }
      return { ...point, color };
    });
  }
  preprocess(data) {
    let workingData = data;
    if (this.options.color === "auto") {
      workingData = this.applyAutoColor(workingData);
      this.applyAutoStackColors(workingData);
      this.options.color = "";
    }
    const sorted = this.sort(workingData);
    const key = this.options.structure.y;
    const total = this.calculateData(workingData);
    const processed = this.options.reverse ? sorted.reverse() : sorted;
    return { processed, key, total };
  }
  process(data) {
    const { processed, total } = this.preprocess(data);
    this.validator.validate(data);
    const chartData = /* @__PURE__ */ new Map();
    processed.forEach((point, index) => {
      chartData.set(index, this.createChartPoint(point, total));
    });
    return [chartData, this.options];
  }
  sort(data) {
    if (!this.options.sort || this.isCandlestick() || this.isStatus()) return data;
    return data.sort((a, b) => this.getPointValue(a) - this.getPointValue(b));
  }
  getSegmentColor(segment, index, pointColors) {
    const segmentColor = typeof segment === "object" ? segment.color : void 0;
    if (segmentColor) return segmentColor;
    if (pointColors?.[index]) return pointColors[index];
    if (this.options.stackColors?.[index]) return this.options.stackColors[index];
    const color = this.options.color;
    return typeof color === "string" ? color : "";
  }
  createChartPoint(point, total) {
    const value = this.getPointValue(point);
    const segments = this.getPointSegments(point, value);
    const ohlc = this.isCandlestick() ? this.getOHLC(point) : void 0;
    const statuses = this.isStatus() ? this.getStatuses(point) : void 0;
    const status = this.isStatus() && !statuses ? this.getStatus(point) : void 0;
    return {
      label: this.getPointLabel(point, value),
      value,
      color: this.getPointColor(point),
      scaled: Number(this.scale(value).toFixed(2)),
      percentage: this.percentage(value, total),
      ...segments && { segments },
      ...ohlc && { ohlc },
      ...status && { status },
      ...statuses && { statuses }
    };
  }
};
var processor_default = ChartProcessor;

// options/options.ts
var defaultOptions = {
  percentage: false,
  colorLabels: true,
  sort: false,
  reverse: false,
  color: void 0,
  title: "",
  labels: true,
  char: "\u2588",
  naked: false,
  width: 100,
  height: 10,
  padding: 0,
  orientation: "horizontal",
  theme: "",
  scale: "auto",
  structure: {
    x: "\u2550",
    y: "\u2562",
    axis: "\u2551",
    topLeft: "\u2554",
    bottomLeft: "\u255A"
  },
  stackColors: void 0,
  stackLabels: void 0,
  stackValueLabels: false,
  alignBars: "justify",
  fillColor: void 0,
  richLabels: true,
  type: "bar",
  variant: "sharp"
};
function createOptions(options) {
  return {
    ...defaultOptions,
    ...options,
    max: {
      label: 0,
      value: 0,
      scaled: 0
    },
    structure: {
      ...defaultOptions.structure,
      ...options?.structure
    }
  };
}

// formatters/vertical.ts
var VerticalChartFormatter = class extends formatter_default {
  constructor(options) {
    super(options);
  }
  calculateDefaultDimensions(barCount, totalSize, charWidth) {
    const defaultBarSize = this.options.barSize || 1;
    const calculatedBarWidth = Math.floor(totalSize / (defaultBarSize * barCount) / charWidth) + 1;
    const barSize = this.options.barSize === void 0 ? calculatedBarWidth : this.options.barSize;
    const calculatedPadding = Math.round(totalSize / barCount / charWidth);
    const defaultPadding = calculatedPadding <= barSize ? 0 : calculatedPadding - barSize;
    const padding = this.options.padding || defaultPadding;
    return { barSize, padding };
  }
  format(chartData) {
    const chart = [...chartData.values()];
    const maxHeight = this.getMaxHeight();
    const charWidth = this.options.char.length;
    const { barSize, padding } = this.calculateBarDimensions(
      chart.length,
      this.options.width,
      charWidth
    );
    const verticalChart = this.buildVerticalChart(chart.length, maxHeight, padding);
    this.formatChart(chart, verticalChart, maxHeight, padding, barSize);
    return this.composeFinalChart(chart, verticalChart, barSize, padding, maxHeight);
  }
  getMaxHeight() {
    return this.options.height + ((this.options.valueLabels || this.options.stackValueLabels) && !this.options.fill ? 1 : 0);
  }
  isLongChar() {
    return this.options.char.length > 1;
  }
  isFillLonger() {
    const length = this.options.fill?.length || 0;
    return length > 0 && length > this.options.char.length;
  }
  getFillChar() {
    const { char, fill } = this.getCharLengths();
    return fill > 0 && fill < char ? this.options.fill.repeat(char) : this.options.fill;
  }
  getCharLengths() {
    return {
      char: this.options.char.length,
      fill: this.options.fill?.length || 0
    };
  }
  getCharWidth() {
    if (this.isLongChar()) return this.options.char.length;
    if (this.isFillLonger()) return this.options.fill.length;
    return 1;
  }
  getScaledBarSize(barSize) {
    const { char, fill } = this.getCharLengths();
    if (fill > 1 && char > 1) return barSize;
    if (this.isFillLonger()) return Math.round(barSize / fill);
    return barSize;
  }
  getHorizontalAlignmentPadding(chartLength, barSize, padding) {
    const alignment = this.options.alignBars || "justify";
    if (alignment === "justify") {
      return { leftPad: 0, rightPad: 0 };
    }
    const charWidth = this.getCharWidth();
    const scaledBarSize = this.getScaledBarSize(barSize);
    const barsWidth = (scaledBarSize * charWidth + padding) * chartLength - padding;
    const { leadingPad, trailingPad } = this.calculateAlignmentPadding(
      barsWidth,
      this.options.width,
      alignment
    );
    return { leftPad: leadingPad, rightPad: trailingPad };
  }
  buildVerticalChart(chartLength, maxHeight, padding) {
    return Array.from(
      { length: maxHeight },
      () => Array.from({ length: chartLength }, () => " ".repeat(padding))
    );
  }
  formatChart(chart, verticalChart, maxHeight, padding, barSize) {
    chart.forEach((point, index) => {
      const isLast = index === chart.length - 1;
      const context = { chart, verticalChart, point, index, maxHeight, padding: isLast ? 0 : padding, barSize };
      if (point.segments?.length) {
        this.formatStackedColumn(context);
      } else {
        this.formatSingleColumn(context);
      }
    });
  }
  formatSingleColumn(ctx) {
    const barHeight = this.calculateBarHeight(ctx.point.scaled, ctx.maxHeight);
    const barStartRow = ctx.maxHeight - barHeight;
    Array.from({ length: ctx.maxHeight }, (_, row) => {
      ctx.verticalChart[row][ctx.index] = this.getCellContent(ctx, row, barStartRow);
    });
  }
  getCellContent(ctx, row, barStartRow) {
    const isValueLabelRow = row === barStartRow - 1 && (this.options.valueLabels || this.options.stackValueLabels) && !this.options.fill;
    const isAboveBar = row < barStartRow;
    const isBarRow = row >= barStartRow;
    if (isValueLabelRow) {
      return this.formatValueLabelCell(ctx.point, ctx.barSize, ctx.padding, ctx.index, ctx.chart.length, ctx.chart, ctx.maxHeight);
    }
    if (isAboveBar) {
      return this.formatEmptyCell(ctx.barSize, ctx.padding, ctx.point.color, row, ctx.index, ctx.chart.length, ctx.maxHeight);
    }
    if (isBarRow) {
      return this.formatBarCellWithContext(ctx, row, barStartRow);
    }
    return this.formatSpace(ctx.barSize, ctx.padding);
  }
  formatBarCellWithContext(ctx, row, barStartRow) {
    const barWidth = this.calculateBarWidth(ctx.barSize);
    const value = this.options.char.repeat(barWidth);
    const padding = " ".repeat(ctx.padding);
    const effectiveColor = ctx.point.color || this.options.color;
    if (this.isGradient(effectiveColor)) {
      const gradient = this.normalizeGradient(effectiveColor);
      const direction = gradient.direction || "horizontal";
      const reverse = gradient.reverse || false;
      const totalBars = ctx.chart.length;
      if (direction === "vertical") {
        const chars = [...value];
        let result = "";
        const charsPerRow = barWidth * totalBars;
        for (let i = 0; i < chars.length; i++) {
          const rowPosition = ctx.maxHeight > 1 ? row / (ctx.maxHeight - 1) : 0;
          const rowSpan = ctx.maxHeight > 1 ? 1 / (ctx.maxHeight - 1) : 1;
          const charIndexInRow = ctx.index * barWidth + i;
          const charFraction = charsPerRow > 1 ? charIndexInRow / (charsPerRow - 1) : 0;
          let position = rowPosition + (charFraction - 0.5) * rowSpan * 0.5;
          position = Math.max(0, Math.min(1, position));
          if (reverse) position = 1 - position;
          const [r, g, b] = this.getColorAtPosition(gradient, position);
          result += `\x1B[38;2;${r};${g};${b}m${chars[i]}\x1B[39m`;
        }
        return result + padding;
      } else if (direction === "diagonal") {
        const chars = [...value];
        let result = "";
        const totalChars = barWidth * totalBars;
        for (let i = 0; i < chars.length; i++) {
          const charIndex = ctx.index * barWidth + i;
          const hPos = totalChars > 1 ? charIndex / (totalChars - 1) : 0;
          const vPos = ctx.maxHeight > 1 ? row / (ctx.maxHeight - 1) : 0;
          let position = (hPos + vPos) / 2;
          if (reverse) position = 1 - position;
          const [r, g, b] = this.getColorAtPosition(gradient, position);
          result += `\x1B[38;2;${r};${g};${b}m${chars[i]}\x1B[39m`;
        }
        return result + padding;
      } else {
        const totalChars = barWidth * totalBars;
        const charIndex = ctx.index * barWidth;
        return this.applyGradientWithContext(value, gradient, charIndex, totalChars, ctx.index, totalBars) + padding;
      }
    }
    return effectiveColor ? this.colorify(value, effectiveColor) + padding : value + padding;
  }
  formatStackedColumn(ctx) {
    const totalHeight = this.calculateBarHeight(ctx.point.scaled, ctx.maxHeight);
    const segments = ctx.point.segments;
    const segmentHeights = segments.map((seg) => this.calculateBarHeight(seg.scaled, ctx.maxHeight));
    this.fillStackedSegments(ctx, segments, segmentHeights);
    this.fillEmptyRows(ctx, totalHeight, segments[0]?.color || ctx.point.color);
    this.addValueLabelIfNeeded(ctx, totalHeight);
  }
  fillStackedSegments(ctx, segments, segmentHeights) {
    let currentRow = ctx.maxHeight - 1;
    segments.forEach((segment, segIdx) => {
      const segmentEndRow = currentRow - segmentHeights[segIdx] + 1;
      for (let row = currentRow; row >= segmentEndRow && row >= 0; row--) {
        ctx.verticalChart[row][ctx.index] = this.formatBarCell(ctx.barSize, ctx.padding, segment.color);
      }
      currentRow = segmentEndRow - 1;
    });
  }
  fillEmptyRows(ctx, totalHeight, color) {
    const emptyStartRow = ctx.maxHeight - totalHeight - 1;
    for (let row = emptyStartRow; row >= 0; row--) {
      ctx.verticalChart[row][ctx.index] = this.formatEmptyCell(ctx.barSize, ctx.padding, color, row, ctx.index, ctx.chart.length, ctx.maxHeight);
    }
  }
  addValueLabelIfNeeded(ctx, totalHeight) {
    const emptyStartRow = ctx.maxHeight - totalHeight - 1;
    const shouldShowLabel = (this.options.valueLabels || this.options.stackValueLabels) && !this.options.fill && emptyStartRow >= 0;
    if (shouldShowLabel) {
      ctx.verticalChart[emptyStartRow][ctx.index] = this.formatValueLabelCell(ctx.point, ctx.barSize, ctx.padding, ctx.index, ctx.chart.length, ctx.chart, ctx.maxHeight);
    }
  }
  calculateBarHeight(scaled, maxHeight) {
    return Math.round(scaled / maxHeight * maxHeight);
  }
  formatValueLabelCell(point, barSize, padding, barIndex, totalBars, chart, maxHeight) {
    const label = this.formatValueLabel(point, barIndex, totalBars, chart, maxHeight);
    const space = Math.max(0, barSize - this.stripStyle(label).length + padding);
    return label + " ".repeat(space);
  }
  formatEmptyCell(barSize, padding, color, row, barIndex, totalBars, maxHeight) {
    return this.options.fill ? this.formatFillCell(barSize, padding, color, row, barIndex, totalBars, maxHeight) : this.formatSpace(barSize, padding);
  }
  formatSpace(barSize, padding) {
    const width = this.isLongChar() ? barSize * this.options.char.length : barSize;
    return " ".repeat(width) + " ".repeat(padding);
  }
  calculateBarWidth(barSize) {
    return this.isFillLonger() ? barSize + (this.options.fill.length - this.options.char.length) : this.getScaledBarSize(barSize);
  }
  formatBarCell(barSize, padding, color) {
    const barWidth = this.calculateBarWidth(barSize);
    const value = this.options.char.repeat(barWidth) + " ".repeat(padding);
    const effectiveColor = color || this.options.color;
    return effectiveColor ? this.colorify(value, effectiveColor) : value;
  }
  formatFillCell(barSize, padding, color, row, barIndex, totalBars, maxHeight) {
    const character = this.getFillChar();
    if (!character) return "";
    const barWidth = this.getScaledBarSize(barSize);
    const paddingStr = " ".repeat(padding);
    const effectiveColor = color || this.options.color;
    if (this.options.fillColor === "auto" && this.isGradient(effectiveColor)) {
      const gradient = this.normalizeGradient(effectiveColor);
      const direction = gradient.direction || "horizontal";
      const reverse = gradient.reverse || false;
      const finalMaxHeight = this.options._finalMaxBarLength ?? maxHeight ?? this.options.height;
      let result = "";
      for (let i = 0; i < barWidth; i++) {
        let position;
        if (direction === "vertical") {
          position = finalMaxHeight > 1 ? (row ?? 0) / (finalMaxHeight - 1) : 0;
        } else if (direction === "diagonal") {
          const charIndex = (barIndex ?? 0) * barWidth + i;
          const totalChars = barWidth * (totalBars ?? 1);
          const hPos = totalChars > 1 ? charIndex / (totalChars - 1) : 0;
          const vPos = finalMaxHeight > 1 ? (row ?? 0) / (finalMaxHeight - 1) : 0;
          position = (hPos + vPos) / 2;
        } else {
          const charIndex = (barIndex ?? 0) * barWidth + i;
          const totalChars = barWidth * (totalBars ?? 1);
          position = totalChars > 1 ? charIndex / (totalChars - 1) : 0;
        }
        if (reverse) position = 1 - position;
        const [r, g, b] = this.getColorAtPosition(gradient, position);
        result += `\x1B[38;2;${r};${g};${b}m${character}\x1B[39m`;
      }
      return result + paddingStr;
    }
    const value = character.repeat(barWidth) + paddingStr;
    let fillColor;
    if (this.options.fillColor === "auto") {
      fillColor = effectiveColor;
    } else if (this.options.fillColor) {
      fillColor = this.options.fillColor;
    } else {
      fillColor = color;
    }
    return fillColor ? this.colorify(value, fillColor) : value;
  }
  formatLabel(point, barIndex, totalBars) {
    const baseLabel = point.percentage ? `${point.label} ${this.formatPercentage(point)}` : point.label;
    const formatted = this.options.labelFormat ? this.options.labelFormat(baseLabel) : baseLabel;
    const label = this.applyDecorators(formatted);
    if (!this.options.colorLabels) return label;
    const pointColor = point.color || this.options.color;
    const labelColor = this.getLabelColorForBar(pointColor, barIndex, totalBars);
    return labelColor ? this.colorify(label, labelColor) : label;
  }
  formatValueLabel(point, barIndex, totalBars, chart, maxHeight) {
    const value = point.segments ? this.formatStackedValueLabel(point.segments) : this.formatValueWithDecimals(point.value);
    if (!this.options.colorLabels) return value;
    const pointColor = point.color || this.options.color;
    let valueLabelColor;
    if (this.options.fill && this.options.fillColor) {
      if (this.options.fillColor === "auto") {
        valueLabelColor = this.getValueLabelColorForBar(pointColor, barIndex, totalBars, 0);
      } else {
        valueLabelColor = this.options.fillColor;
      }
    } else {
      let barEndPosition;
      if (chart && maxHeight && this.isGradient(pointColor)) {
        const gradient = this.normalizeGradient(pointColor);
        if (gradient.direction === "vertical") {
          const barHeight = this.calculateBarHeight(point.scaled, maxHeight);
          const barStartRow = maxHeight - barHeight;
          barEndPosition = maxHeight > 1 ? barStartRow / (maxHeight - 1) : 0;
        }
      }
      valueLabelColor = this.getValueLabelColorForBar(pointColor, barIndex, totalBars, barEndPosition);
    }
    return valueLabelColor ? this.colorify(value, valueLabelColor) : value;
  }
  formatLabels(chart, barSize, padding) {
    if (!this.options.labels) return "";
    const { leftPad } = this.getHorizontalAlignmentPadding(chart.length, barSize, padding);
    const labels = chart.map((point, i) => this.formatLabelEntry(point, barSize, i === chart.length - 1 ? 0 : padding, i, chart.length)).join("");
    return " ".repeat(leftPad) + labels;
  }
  formatLabelEntry(point, barSize, padding, index, totalBars) {
    return this.formatEntryWithLabel(this.formatLabel(point, index, totalBars), barSize, padding, index);
  }
  formatValueLabels(chart, barSize, padding, maxHeight) {
    if (!this.options.labels) return "";
    const { leftPad } = this.getHorizontalAlignmentPadding(chart.length, barSize, padding);
    const labels = chart.map((point, i) => this.formatValueLabelEntry(point, barSize, i === chart.length - 1 ? 0 : padding, i, chart.length, chart, maxHeight)).join("");
    return " ".repeat(leftPad) + labels;
  }
  formatValueLabelEntry(point, barSize, padding, index, totalBars, chart, maxHeight) {
    return this.formatEntryWithLabel(this.formatValueLabel(point, index, totalBars, chart, maxHeight), barSize, padding, index);
  }
  formatEntryWithLabel(formattedLabel, barSize, padding, index) {
    const label = this.stripStyle(formattedLabel);
    const charLength = this.getCharWidth();
    const barWidth = this.isLongChar() ? barSize * charLength + padding : barSize + padding + Math.floor(charLength / 2);
    const rightPad = Math.abs(barWidth - label.length);
    const isFirst = index === 0 && !this.options.naked ? 1 : 0;
    return " ".repeat(isFirst) + formattedLabel + " ".repeat(rightPad);
  }
  composeFinalChart(chart, verticalChart, barSize, padding, maxHeight) {
    const chartRows = this.buildChartRows(chart.length, verticalChart, barSize, padding);
    const header = this.buildHeader(chart.length, barSize, padding);
    const footer = this.buildFooter(chart, barSize, padding);
    const valueLabelsHeader = this.buildValueLabelsHeader(chart, barSize, padding, maxHeight);
    return [...header, ...valueLabelsHeader, ...chartRows, ...footer].join("\n");
  }
  buildChartRows(chartLength, verticalChart, barSize, padding) {
    const { leftPad, rightPad } = this.getHorizontalAlignmentPadding(chartLength, barSize, padding);
    return verticalChart.map((row) => {
      const rowContent = row.join("");
      const alignedContent = " ".repeat(leftPad) + rowContent + " ".repeat(rightPad);
      return this.options.naked ? alignedContent : this.options.structure.axis + alignedContent;
    });
  }
  buildHeader(chartLength, barSize, padding) {
    return this.options.title ? [this.formatChartTitle(chartLength, barSize, padding)] : [];
  }
  buildFooter(chart, barSize, padding) {
    const footer = [];
    if (!this.options.naked) {
      footer.push(this.formatBottom(chart.length, barSize, padding));
    } else if (this.options.labels) {
      footer.push("");
    }
    if (this.options.labels) {
      footer.push(this.formatLabels(chart, barSize, padding));
    }
    return footer;
  }
  buildValueLabelsHeader(chart, barSize, padding, maxHeight) {
    if (!(this.options.valueLabels || this.options.stackValueLabels) || !this.options.fill) return [];
    return [this.formatValueLabels(chart, barSize, padding, maxHeight)];
  }
  formatChartTitle(chartLength, barSize, padding) {
    const alignment = this.options.alignBars || "justify";
    if (alignment === "justify") {
      const charLength = this.getCharWidth();
      const barWidth = this.getScaledBarSize(barSize);
      const barsWidth = (barWidth * charLength + padding) * chartLength - padding;
      return this.formatTitle(barsWidth + 1);
    }
    return this.formatTitle(this.options.width + 1);
  }
  formatBottom(chartLength, barSize, padding) {
    const alignment = this.options.alignBars || "justify";
    const charLength = this.getCharWidth();
    const barWidth = this.getScaledBarSize(barSize);
    const barsWidth = (barWidth * charLength + padding) * chartLength - padding;
    if (alignment === "justify") {
      return this.options.structure.bottomLeft + this.options.structure.x.repeat(barsWidth);
    }
    return this.options.structure.bottomLeft + this.options.structure.x.repeat(this.options.width);
  }
};
var vertical_default = VerticalChartFormatter;

// formatters/line.ts
function labelMinColDiff(L1, L2) {
  if (L1 <= 0 && L2 <= 0) return 1;
  return Math.max(1, Math.ceil(L1 / 2) + Math.floor(L2 / 2) + 1);
}
var LINE_CHARS = {
  flatTop: "\u203E",
  flatBottom: "_",
  up: "\u2571",
  down: "\u2572"
};
var LineChartFormatter = class extends formatter_default {
  constructor(options) {
    super(options);
  }
  calculateDefaultDimensions(barCount, totalSize, charWidth) {
    return { barSize: 1, padding: 0 };
  }
  format(chartData) {
    const chart = [...chartData.values()];
    if (chart.length === 0) return "";
    const height = this.options.height || 10;
    const width = this.options.width || 80;
    const maxValue = this.options.max.value || 1;
    const minValue = this.options.max.min ?? 0;
    const yAxisTicks = this.buildYAxisTicks(minValue, maxValue, height);
    const yLabelWidth = Math.max(...yAxisTicks.map((t) => t.label.length));
    const requestedWidth = width - yLabelWidth - 1;
    if (requestedWidth < 2) return "";
    const points = this.mapPointsToGrid(chart, requestedWidth, height, minValue, maxValue);
    const chartWidth = this.prepareLayout(points, requestedWidth);
    const grid = this.buildGrid(chartWidth, height);
    const seriesColor = this.options._seriesColors?.[0];
    let gridColors;
    let lineBottom;
    if (seriesColor !== void 0) {
      gridColors = Array.from({ length: height }, () => new Array(chartWidth).fill(void 0));
      lineBottom = this.drawLineColored(grid, gridColors, points, chartWidth, seriesColor);
    } else {
      lineBottom = this.drawLine(grid, points, chartWidth);
    }
    if (this.options.fill) {
      this.drawAreaFill(grid, lineBottom, height);
    }
    if (this.shouldDrawPoints()) {
      const pointChar = this.options.pointChar || "\u25CF";
      if (gridColors) {
        this.drawPointsColored(grid, gridColors, points, pointChar, seriesColor);
      } else {
        this.drawPoints(grid, points, pointChar);
      }
    }
    this.applyAnimationClip(grid, gridColors, chartWidth);
    return this.compose(grid, yAxisTicks, yLabelWidth, chartWidth, height, points, gridColors);
  }
  /**
   * Multi-line entry point. Each `ChartData` in `chartDataArr` becomes its
   * own line series, sharing the y-axis scale (global max/min across all
   * series) and x-axis layout (per-segment colDiff = max over series so
   * every series keeps a 45°-continuous line; the series with a smaller
   * row-diff at a segment fills the extra cols with `‾`/`_`). Each series
   * is colored by `options.lineColor[i]`; points (if enabled) inherit the
   * series color.
   */
  formatMulti(chartDataArr) {
    if (chartDataArr.length === 0) return "";
    if (chartDataArr.length === 1) return this.format(chartDataArr[0]);
    const chartArrays = chartDataArr.map((cd) => [...cd.values()]);
    const maxLen = Math.max(...chartArrays.map((a) => a.length));
    if (maxLen === 0) return "";
    const height = this.options.height || 10;
    const width = this.options.width || 80;
    const maxValue = this.options.max.value || 1;
    const minValue = this.options.max.min ?? 0;
    const yAxisTicks = this.buildYAxisTicks(minValue, maxValue, height);
    const yLabelWidth = Math.max(...yAxisTicks.map((t) => t.label.length));
    const requestedWidth = width - yLabelWidth - 1;
    if (requestedWidth < 2) return "";
    const seriesPoints = chartArrays.map(
      (chart) => this.mapPointsToGrid(chart, requestedWidth, height, minValue, maxValue)
    );
    const chartWidth = this.prepareLayoutMulti(seriesPoints, requestedWidth);
    const grid = this.buildGrid(chartWidth, height);
    const gridColors = Array.from(
      { length: height },
      () => new Array(chartWidth).fill(void 0)
    );
    const seriesColors = this.options._seriesColors || [];
    for (let s = 0; s < seriesPoints.length; s++) {
      this.drawLineColored(grid, gridColors, seriesPoints[s], chartWidth, seriesColors[s]);
    }
    if (this.shouldDrawPoints()) {
      const pointChar = this.options.pointChar || "\u25CF";
      for (let s = 0; s < seriesPoints.length; s++) {
        this.drawPointsColored(grid, gridColors, seriesPoints[s], pointChar, seriesColors[s]);
      }
    }
    this.applyAnimationClip(grid, gridColors, chartWidth);
    return this.compose(grid, yAxisTicks, yLabelWidth, chartWidth, height, seriesPoints[0], gridColors);
  }
  /**
   * Whether to render point markers. Subclasses (scatter) can override
   * to render points unconditionally.
   */
  shouldDrawPoints() {
    return Boolean(this.options.points);
  }
  /**
   * Left-to-right reveal for animation. Clears any data-area cells past
   * the visible cutoff so line segments / points beyond the current
   * progress disappear. Only the data grid is touched — y-axis, x-labels,
   * legend, title, and the bottom border still render fully so the layout
   * stays stable across frames.
   */
  applyAnimationClip(grid, gridColors, chartWidth) {
    const progress = this.options._animationProgress;
    if (progress === void 0 || progress >= 1) return;
    const cutoff = Math.max(0, Math.min(chartWidth, Math.ceil(progress * chartWidth)));
    if (cutoff >= chartWidth) return;
    for (let row = 0; row < grid.length; row++) {
      for (let col = cutoff; col < chartWidth; col++) {
        grid[row][col] = " ";
        if (gridColors) gridColors[row][col] = void 0;
      }
    }
  }
  /**
   * Repositions points so each segment's column span is at least its
   * natural 45° width (and the label-spacing minimum), then expands the
   * spacing to fill `requestedWidth` when there's room — `drawLine` emits
   * `‾`/`_` flats to bridge the extra cols inside each segment, so the
   * diagonals stay 45° at peaks/troughs. Returns the effective chart
   * width. Mutates each point's `col` and `anchor` in place. Subclasses
   * (step) may override to drop the 45° constraint entirely.
   */
  prepareLayout(points, requestedWidth) {
    if (points.length === 0) return requestedWidth;
    for (let i = 0; i < points.length; i++) {
      if (i === 0) {
        let firstNonFlat = 0;
        for (let j = 1; j < points.length; j++) {
          const d = points[j].row - points[j - 1].row;
          if (d !== 0) {
            firstNonFlat = d;
            break;
          }
        }
        points[i].anchor = firstNonFlat < 0 ? 1 : 0;
      } else {
        const d = points[i].row - points[i - 1].row;
        if (d < 0) points[i].anchor = 0;
        else if (d > 0) points[i].anchor = 1;
        else points[i].anchor = points[i - 1].anchor;
      }
    }
    const naturalDiffs = this.computeNaturalDiffs(points);
    const firstLabel = points[0].point?.label ?? "";
    const lastLabel = points[points.length - 1].point?.label ?? "";
    const leadPad = Math.floor(firstLabel.length / 2);
    const trailPad = Math.max(0, lastLabel.length - 1 - Math.floor(lastLabel.length / 2));
    const evenSpan = Math.max(0, requestedWidth - 1 - leadPad - trailPad);
    points[0].col = 0;
    for (let i = 1; i < points.length; i++) {
      const evenCol = points.length > 1 ? Math.round(i * evenSpan / (points.length - 1)) : 0;
      const minCol = points[i - 1].col + naturalDiffs[i];
      points[i].col = Math.max(evenCol, minCol);
    }
    if (leadPad > 0) for (const p of points) p.col += leadPad;
    return Math.max(requestedWidth, points[points.length - 1].col + 1 + trailPad);
  }
  /**
   * Per-segment minimum colDiff for a single line series: max of the 45°
   * natural span (or 1 for flat segments), the label-spacing minimum, and
   * — when points are rendered — a -1 peak/trough shrink so the `●` lands
   * diagonally adjacent to the next change instead of two cols away.
   * Index 0 is unused (segments are between points i-1 and i).
   */
  computeNaturalDiffs(points) {
    const diffs = [0];
    for (let i = 1; i < points.length; i++) {
      const yFrom = points[i - 1].row + (points[i - 1].anchor ?? 0);
      const yTo = points[i].row + (points[i].anchor ?? 0);
      const isFlat = points[i].row === points[i - 1].row;
      let colDiff = isFlat ? 1 : Math.max(Math.abs(yTo - yFrom), 1);
      if (this.options.points && !isFlat) {
        const prev = i - 1;
        if (prev > 0 && prev < points.length - 1) {
          const pr = points[prev].row;
          const beforeRow = points[prev - 1].row;
          const afterRow = points[prev + 1].row;
          const isPrevPeak = beforeRow > pr && afterRow > pr;
          const isPrevTrough = beforeRow < pr && afterRow < pr;
          if (isPrevPeak || isPrevTrough) colDiff = Math.max(1, colDiff - 1);
        }
      }
      const L1 = this.labelLength(points[i - 1].point?.label);
      const L2 = this.labelLength(points[i].point?.label);
      const labelMin = labelMinColDiff(L1, L2);
      if (labelMin > colDiff) colDiff = labelMin;
      diffs.push(colDiff);
    }
    return diffs;
  }
  /**
   * Layout for multi-series. Each series still gets its own anchor
   * sequence (its own rising/falling pattern), but every segment's
   * column span is the max over all series — so the widest natural 45°
   * wins, and every series lines up on shared data-point columns.
   * Series with smaller row-diffs at a segment will use Bresenham stays
   * (`‾`/`_`) to fill the extra cols. When the requested width exceeds
   * the natural minimum, the surplus is distributed evenly so the chart
   * fills the requested width.
   */
  prepareLayoutMulti(seriesPoints, requestedWidth) {
    if (seriesPoints.length === 0) return requestedWidth;
    const maxLen = Math.max(...seriesPoints.map((p) => p.length));
    if (maxLen === 0) return requestedWidth;
    for (const points of seriesPoints) {
      for (let i = 0; i < points.length; i++) {
        if (i === 0) {
          let firstNonFlat = 0;
          for (let j = 1; j < points.length; j++) {
            const d = points[j].row - points[j - 1].row;
            if (d !== 0) {
              firstNonFlat = d;
              break;
            }
          }
          points[i].anchor = firstNonFlat < 0 ? 1 : 0;
        } else {
          const d = points[i].row - points[i - 1].row;
          if (d < 0) points[i].anchor = 0;
          else if (d > 0) points[i].anchor = 1;
          else points[i].anchor = points[i - 1].anchor;
        }
      }
    }
    const naturalDiffs = [0];
    const firstSeries = seriesPoints[0];
    for (let i = 1; i < maxLen; i++) {
      let maxColDiff = 1;
      const L1 = this.labelLength(firstSeries[i - 1]?.point?.label);
      const L2 = this.labelLength(firstSeries[i]?.point?.label);
      const labelMin = labelMinColDiff(L1, L2);
      for (const points of seriesPoints) {
        if (i >= points.length) continue;
        const yFrom = points[i - 1].row + (points[i - 1].anchor ?? 0);
        const yTo = points[i].row + (points[i].anchor ?? 0);
        const isFlat = points[i].row === points[i - 1].row;
        const natural = isFlat ? 1 : Math.max(Math.abs(yTo - yFrom), 1);
        let colDiff = natural;
        if (this.options.points && !isFlat) {
          const prev = i - 1;
          if (prev > 0 && prev < points.length - 1) {
            const pr = points[prev].row;
            const beforeRow = points[prev - 1].row;
            const afterRow = points[prev + 1].row;
            if (beforeRow > pr && afterRow > pr || beforeRow < pr && afterRow < pr) {
              colDiff = Math.max(1, colDiff - 1);
            }
          }
        }
        if (colDiff > maxColDiff) maxColDiff = colDiff;
      }
      if (labelMin > maxColDiff) maxColDiff = labelMin;
      naturalDiffs.push(maxColDiff);
    }
    const firstLabel = firstSeries[0]?.point?.label ?? "";
    const lastLabel = firstSeries[firstSeries.length - 1]?.point?.label ?? "";
    const leadPad = Math.floor(firstLabel.length / 2);
    const trailPad = Math.max(0, lastLabel.length - 1 - Math.floor(lastLabel.length / 2));
    const evenSpan = Math.max(0, requestedWidth - 1 - leadPad - trailPad);
    const sharedCols = [0];
    for (let i = 1; i < maxLen; i++) {
      const evenCol = maxLen > 1 ? Math.round(i * evenSpan / (maxLen - 1)) : 0;
      const minCol = sharedCols[i - 1] + naturalDiffs[i];
      sharedCols.push(Math.max(evenCol, minCol));
    }
    for (const points of seriesPoints) {
      for (let i = 0; i < points.length; i++) {
        if (i < sharedCols.length) points[i].col = sharedCols[i];
      }
    }
    if (leadPad > 0) {
      for (const points of seriesPoints) {
        for (const p of points) p.col += leadPad;
      }
    }
    const lastCol = Math.max(
      ...seriesPoints.map((p) => p[p.length - 1]?.col ?? 0)
    );
    return Math.max(requestedWidth, lastCol + 1 + trailPad);
  }
  /**
   * Draws each segment as pure 45° diagonals using a state machine over
   * (row, pos). `╱` moves the line up one half-step; `╲` moves down; `‾`
   * or `_` keeps the line on its current anchor. When `colDiff > absYDiff`
   * the row changes are placed contiguously (clustered) rather than
   * spread evenly — see `pickClusterAnchor` — so slopes look like a
   * single straight diagonal with a flat plateau on one side, rather
   * than a jagged staircase.
   */
  drawLine(grid, points, chartWidth) {
    const chars = LINE_CHARS;
    const lineBottom = new Array(chartWidth).fill(-1);
    if (points.length === 0) return lineBottom;
    const put = (row, col, ch) => {
      if (row >= 0 && row < grid.length && col >= 0 && col < chartWidth) {
        grid[row][col] = ch;
        lineBottom[col] = row;
      }
    };
    for (let i = 0; i < points.length - 1; i++) {
      const p1 = points[i];
      const p2 = points[i + 1];
      const colDiff = p2.col - p1.col;
      if (colDiff <= 0) continue;
      let p1Override = null;
      const rowDiff = p2.row - p1.row;
      if (this.options.points && rowDiff !== 0 && i > 0 && i < points.length - 1) {
        const beforeRow = points[i - 1].row;
        const afterRow = points[i + 1].row;
        const pr = p1.row;
        if (beforeRow > pr && afterRow > pr) p1Override = 1;
        else if (beforeRow < pr && afterRow < pr) p1Override = 0;
      }
      const startPos = p1Override ?? (p1.anchor ?? 0);
      const yStart = p1.row + startPos;
      const yEnd = p2.row + (p2.anchor ?? 0);
      const yDiff = yEnd - yStart;
      const absYDiff = Math.abs(yDiff);
      const changeCols = this.buildChangeCols(points, i, p1.col, p2.col, absYDiff, colDiff);
      let row = p1.row;
      let pos = startPos;
      for (let col = p1.col + 1; col <= p2.col; col++) {
        if (changeCols.has(col)) {
          if (yDiff < 0) {
            if (pos === 1) {
              put(row, col, chars.up);
              pos = 0;
            } else {
              row -= 1;
              put(row, col, chars.up);
            }
          } else {
            if (pos === 0) {
              put(row, col, chars.down);
              pos = 1;
            } else {
              row += 1;
              put(row, col, chars.down);
            }
          }
        } else {
          put(row, col, pos === 0 ? chars.flatTop : chars.flatBottom);
        }
      }
    }
    return lineBottom;
  }
  /**
   * Decides whether a segment's row changes should cluster at the start
   * (`╱╱╱‾‾‾`, line leaves p_a sharply) or end (`‾‾‾╱╱╱`, line lands on
   * p_b sharply). The rule:
   *   - If p_b is a peak/trough/last point → end (sharp arrival)
   *   - Else if p_a is a peak/trough/first point → start (sharp departure)
   *   - Otherwise (both endpoints are continuing) → end
   * This keeps peaks/troughs sharp `‾‾╱╲‾‾` and avoids the jagged
   * `‾╱‾╱‾╱` staircase that even-spread Bresenham produces when the
   * segment's colDiff exceeds its rowDiff.
   */
  pickClusterAnchor(points, i) {
    if (i + 1 === points.length - 1) return "end";
    const pbRow = points[i + 1].row;
    const pbBefore = points[i].row;
    const pbAfter = points[i + 2].row;
    if (pbBefore > pbRow && pbAfter > pbRow || pbBefore < pbRow && pbAfter < pbRow) {
      return "end";
    }
    if (i === 0) return "start";
    const paRow = points[i].row;
    const paBefore = points[i - 1].row;
    const paAfter = points[i + 1].row;
    if (paBefore > paRow && paAfter > paRow || paBefore < paRow && paAfter < paRow) {
      return "start";
    }
    return "end";
  }
  /**
   * Builds the set of columns where the line drawing should emit a row
   * change (`╱` or `╲`). Cols not in the set get a flat `‾`/`_` filler.
   * Changes are placed contiguously at the segment's start or end based
   * on `pickClusterAnchor`.
   */
  buildChangeCols(points, i, p1Col, p2Col, absYDiff, colDiff) {
    const changeCols = /* @__PURE__ */ new Set();
    if (absYDiff <= 0) return changeCols;
    const steps = Math.min(absYDiff, colDiff);
    const anchor = this.pickClusterAnchor(points, i);
    if (anchor === "start") {
      for (let k = 1; k <= steps; k++) changeCols.add(p1Col + k);
    } else {
      for (let k = 0; k < steps; k++) changeCols.add(p2Col - k);
    }
    return changeCols;
  }
  buildYAxisTicks(min, max, height) {
    const range = max - min || 1;
    const rawStep = range / 5;
    const magnitude = Math.pow(10, Math.floor(Math.log10(rawStep)));
    const niceSteps = [1, 2, 5, 10];
    const step = niceSteps.find((s) => s * magnitude >= rawStep) * magnitude;
    const tickMin = Math.floor(min / step) * step;
    const tickMax = Math.ceil(max / step) * step;
    const ticks = [];
    for (let value = tickMin; value <= tickMax; value += step) {
      const normalized = (value - min) / range;
      const row = height - 1 - Math.round(normalized * (height - 1));
      if (row < 0 || row >= height) continue;
      const label = Number.isInteger(value) ? String(value) : value.toFixed(1);
      ticks.push({ label, row });
    }
    if (!ticks.some((t) => t.label === String(min))) {
      const label = Number.isInteger(min) ? String(min) : min.toFixed(1);
      ticks.push({ label, row: height - 1 });
    }
    if (!ticks.some((t) => t.label === String(max))) {
      const label = Number.isInteger(max) ? String(max) : max.toFixed(1);
      ticks.push({ label, row: 0 });
    }
    ticks.sort((a, b) => a.row - b.row);
    const seen = /* @__PURE__ */ new Set();
    return ticks.filter((t) => {
      if (seen.has(t.row)) return false;
      seen.add(t.row);
      return true;
    });
  }
  /**
   * Visual length of a label, stripping rich-text decorator markers
   * (`*bold*`, `%italic%`, etc.) which become invisible ANSI codes.
   */
  labelLength(label) {
    if (!label) return 0;
    if (this.options.richLabels === false) return label.length;
    return label.replace(/([*~%!^@#$])(.+?)\1/g, "$2").length;
  }
  mapPointsToGrid(chart, chartWidth, height, minValue, maxValue) {
    const range = maxValue - minValue || 1;
    return chart.map((point, index) => {
      const col = chart.length > 1 ? Math.round(index / (chart.length - 1) * (chartWidth - 1)) : Math.floor(chartWidth / 2);
      const normalizedValue = (point.value - minValue) / range;
      const row = height - 1 - Math.round(normalizedValue * (height - 1));
      return { col, row: Math.max(0, Math.min(height - 1, row)), point };
    });
  }
  buildGrid(width, height) {
    return Array.from(
      { length: height },
      () => Array.from({ length: width }, () => " ")
    );
  }
  drawAreaFill(grid, lineBottom, height) {
    const fillChar = this.options.fill;
    for (let col = 0; col < lineBottom.length; col++) {
      const lineRow = lineBottom[col];
      if (lineRow < 0) continue;
      for (let row = lineRow + 1; row < height; row++) {
        if (grid[row][col] === " ") {
          grid[row][col] = fillChar;
        }
      }
    }
  }
  drawPoints(grid, points, char) {
    for (const { col, row } of points) {
      if (row >= 0 && row < grid.length && col >= 0 && col < grid[0].length) {
        grid[row][col] = char;
        const nextCol = col + 1;
        if (nextCol < grid[row].length) {
          const nextCh = grid[row][nextCol];
          if (nextCh === "\u2571" || nextCh === "\u2572" || nextCh === "\u203E" || nextCh === "_") {
            grid[row][nextCol] = " ";
          }
        }
      }
    }
  }
  /**
   * Resolve the cell color at column `col` for a given series color and
   * series x-extent. Returns a flat color string for plain colors, or an
   * RGB-encoded ANSI escape string interpolated from a gradient.
   *
   * For gradients, the position is the cell's normalized progress between
   * the series' first and last cols (`direction: 'vertical'` uses the row
   * within the chart height instead).
   */
  resolveCellColor(color, col, row, seriesStart, seriesEnd, height, reverseDefault = false) {
    if (color === void 0 || color === "") return void 0;
    const normalized = typeof color === "string" ? normalizeColor(color) : color;
    if (!isGradientObject(normalized)) {
      return normalized;
    }
    const grad = normalized;
    const { direction = "horizontal", reverse = false } = grad;
    let position;
    if (direction === "vertical") {
      position = height > 1 ? row / (height - 1) : 0;
    } else if (direction === "diagonal") {
      const span = seriesEnd - seriesStart;
      const hPos = span > 0 ? (col - seriesStart) / span : 0;
      const vPos = height > 1 ? row / (height - 1) : 0;
      position = (hPos + vPos) / 2;
    } else {
      const span = seriesEnd - seriesStart;
      position = span > 0 ? (col - seriesStart) / span : 0;
    }
    if (reverse !== reverseDefault) position = 1 - position;
    const [r, g, b] = this.getColorAtPosition(grad, position);
    return `\x1B[38;2;${r};${g};${b}m`;
  }
  /** Variant of drawLine that also writes per-cell colors (for multi-series). */
  drawLineColored(grid, gridColors, points, chartWidth, color) {
    const chars = LINE_CHARS;
    const lineBottom = new Array(chartWidth).fill(-1);
    if (points.length === 0) return lineBottom;
    const seriesStart = points[0].col;
    const seriesEnd = points[points.length - 1].col;
    const height = grid.length;
    const put = (row, col, ch) => {
      if (row >= 0 && row < grid.length && col >= 0 && col < chartWidth) {
        grid[row][col] = ch;
        const cellColor = this.resolveCellColor(color, col, row, seriesStart, seriesEnd, height);
        if (cellColor) gridColors[row][col] = cellColor;
        lineBottom[col] = row;
      }
    };
    for (let i = 0; i < points.length - 1; i++) {
      const p1 = points[i];
      const p2 = points[i + 1];
      const colDiff = p2.col - p1.col;
      if (colDiff <= 0) continue;
      let p1Override = null;
      const rowDiff = p2.row - p1.row;
      if (this.options.points && rowDiff !== 0 && i > 0 && i < points.length - 1) {
        const beforeRow = points[i - 1].row;
        const afterRow = points[i + 1].row;
        const pr = p1.row;
        if (beforeRow > pr && afterRow > pr) p1Override = 1;
        else if (beforeRow < pr && afterRow < pr) p1Override = 0;
      }
      const startPos = p1Override ?? (p1.anchor ?? 0);
      const yStart = p1.row + startPos;
      const yEnd = p2.row + (p2.anchor ?? 0);
      const yDiff = yEnd - yStart;
      const absYDiff = Math.abs(yDiff);
      const changeCols = this.buildChangeCols(points, i, p1.col, p2.col, absYDiff, colDiff);
      let row = p1.row;
      let pos = startPos;
      for (let col = p1.col + 1; col <= p2.col; col++) {
        if (changeCols.has(col)) {
          if (yDiff < 0) {
            if (pos === 1) {
              put(row, col, chars.up);
              pos = 0;
            } else {
              row -= 1;
              put(row, col, chars.up);
            }
          } else {
            if (pos === 0) {
              put(row, col, chars.down);
              pos = 1;
            } else {
              row += 1;
              put(row, col, chars.down);
            }
          }
        } else {
          put(row, col, pos === 0 ? chars.flatTop : chars.flatBottom);
        }
      }
    }
    return lineBottom;
  }
  /** Variant of drawPoints that also writes per-cell colors (for multi-series). */
  drawPointsColored(grid, gridColors, points, char, color) {
    if (points.length === 0) return;
    const seriesStart = points[0].col;
    const seriesEnd = points[points.length - 1].col;
    const height = grid.length;
    for (const { col, row } of points) {
      if (row >= 0 && row < grid.length && col >= 0 && col < grid[0].length) {
        grid[row][col] = char;
        const cellColor = this.resolveCellColor(color, col, row, seriesStart, seriesEnd, height);
        if (cellColor) gridColors[row][col] = cellColor;
        const nextCol = col + 1;
        if (nextCol < grid[row].length) {
          const nextCh = grid[row][nextCol];
          if (nextCh === "\u2571" || nextCh === "\u2572" || nextCh === "\u203E" || nextCh === "_") {
            grid[row][nextCol] = " ";
            gridColors[row][nextCol] = void 0;
          }
        }
      }
    }
  }
  compose(grid, yAxisTicks, yLabelWidth, chartWidth, height, points, gridColors) {
    const lines = [];
    const isNaked = this.options.naked;
    const axisChar = isNaked ? " " : this.options.structure.axis;
    const title = this.formatTitle(yLabelWidth + 1 + chartWidth);
    if (title) lines.push(title);
    const legendConfig = this.resolveLegendConfig();
    const legendRows = legendConfig ? this.buildLegendRows(legendConfig, chartWidth) : [];
    const legendPrefix = " ".repeat(yLabelWidth) + axisChar;
    if (legendConfig?.position === "top") {
      for (const row of legendRows) lines.push(legendPrefix + row);
      lines.push(legendPrefix + " ".repeat(chartWidth));
    }
    for (let row = 0; row < height; row++) {
      const tick = yAxisTicks.find((t) => t.row === row);
      const yLabel = tick ? tick.label.padStart(yLabelWidth) : " ".repeat(yLabelWidth);
      const rowContent = this.colorizeRow(grid[row], row, height, chartWidth, gridColors?.[row]);
      lines.push(yLabel + axisChar + rowContent);
    }
    if (legendConfig?.position === "bottom") {
      lines.push(legendPrefix + " ".repeat(chartWidth));
      for (const row of legendRows) lines.push(legendPrefix + row);
    }
    if (!isNaked) {
      lines.push(
        " ".repeat(yLabelWidth) + this.options.structure.bottomLeft + this.options.structure.x.repeat(chartWidth)
      );
    }
    if (this.options.labels !== false) {
      lines.push(this.formatXLabels(points, yLabelWidth + 1, chartWidth));
    }
    return lines.join("\n");
  }
  colorizeRow(row, rowIndex, height, width, perCellColors) {
    if (perCellColors) {
      return row.map((char, colIndex) => {
        if (char === " ") return char;
        const cellColor = perCellColors[colIndex];
        if (!cellColor) return char;
        if (cellColor.startsWith("\x1B[38;2;")) {
          return `${cellColor}${char}\x1B[39m`;
        }
        return this.colorify(char, cellColor);
      }).join("");
    }
    const effectiveColor = this.options.color;
    if (!effectiveColor) return row.join("");
    const normalized = typeof effectiveColor === "string" ? normalizeColor(effectiveColor) : effectiveColor;
    if (isGradientObject(normalized)) {
      return this.colorizeRowGradient(row, rowIndex, height, width, normalized);
    }
    return row.map((char) => {
      if (char === " ") return char;
      return this.colorify(char, effectiveColor);
    }).join("");
  }
  colorizeRowGradient(row, rowIndex, height, width, gradient) {
    const { direction = "horizontal", reverse = false } = gradient;
    return row.map((char, colIndex) => {
      if (char === " ") return char;
      let position;
      if (direction === "vertical") {
        position = height > 1 ? rowIndex / (height - 1) : 0;
      } else if (direction === "diagonal") {
        const hPos = width > 1 ? colIndex / (width - 1) : 0;
        const vPos = height > 1 ? rowIndex / (height - 1) : 0;
        position = (hPos + vPos) / 2;
      } else {
        position = width > 1 ? colIndex / (width - 1) : 0;
      }
      if (reverse) position = 1 - position;
      const [r, g, b] = this.getColorAtPosition(gradient, position);
      return `\x1B[38;2;${r};${g};${b}m${char}\x1B[39m`;
    }).join("");
  }
  formatXLabels(points, offset, chartWidth) {
    const labelLine = Array.from({ length: chartWidth }, () => " ");
    for (const { col, point } of points) {
      const label = point.label;
      const start = Math.max(0, col - Math.floor(label.length / 2));
      for (let i = 0; i < label.length && start + i < chartWidth; i++) {
        labelLine[start + i] = label[i];
      }
    }
    return " ".repeat(offset) + labelLine.join("");
  }
  /**
   * Normalize the user-facing `legend` option. Returns `undefined` when the
   * legend is disabled or the chart is single-series (multi-series only —
   * a single-entry legend would be useless).
   */
  resolveLegendConfig() {
    const seriesCount = this.options._seriesColors?.length ?? 0;
    if (seriesCount <= 1) return void 0;
    const raw = this.options.legend;
    if (!raw) return void 0;
    const config = raw === true ? {} : raw;
    if (config.enabled === false) return void 0;
    return {
      enabled: true,
      values: config.values ?? [],
      position: config.position ?? "top",
      align: config.align ?? "left"
    };
  }
  /**
   * Render legend entries into `chartWidth`-wide rows. Each entry uses the
   * series color as background with an auto-contrasted (luminance-based)
   * foreground. Greedy-wraps to a new row when an entry would overflow,
   * with a single space separator between entries on the same row.
   */
  buildLegendRows(config, chartWidth) {
    const seriesColors = this.options._seriesColors ?? [];
    const labels = seriesColors.map((_, i) => config.values[i] ?? `Series #${i + 1}`);
    const sharedGradient = this.sharedSeriesGradient(seriesColors);
    if (sharedGradient) {
      return this.buildGradientLegendRows(labels, sharedGradient, config, chartWidth);
    }
    const entries = labels.map((label, i) => {
      const bgRgb = this.resolveLegendSwatchRgb(seriesColors[i]);
      if (!bgRgb) return { rendered: label, visualWidth: label.length };
      const [fr, fg, fb] = this.contrastFgRgb(bgRgb);
      const [br, bg, bb] = bgRgb;
      const rendered = `\x1B[48;2;${br};${bg};${bb}m\x1B[38;2;${fr};${fg};${fb}m${label}\x1B[39m\x1B[49m`;
      return { rendered, visualWidth: label.length };
    });
    const rows = [[]];
    let currentWidth = 0;
    for (const entry of entries) {
      const sepWidth = rows[rows.length - 1].length > 0 ? 1 : 0;
      const additional = entry.visualWidth + sepWidth;
      if (currentWidth + additional > chartWidth && rows[rows.length - 1].length > 0) {
        rows.push([entry]);
        currentWidth = entry.visualWidth;
      } else {
        rows[rows.length - 1].push(entry);
        currentWidth += additional;
      }
    }
    return rows.filter((row) => row.length > 0).map((row) => {
      const visualWidth = row.reduce(
        (sum, e, i) => sum + e.visualWidth + (i > 0 ? 1 : 0),
        0
      );
      const content = row.map((e) => e.rendered).join(" ");
      const pad = Math.max(0, chartWidth - visualWidth);
      if (config.align === "right") return " ".repeat(pad) + content;
      if (config.align === "center") {
        const left = Math.floor(pad / 2);
        return " ".repeat(left) + content + " ".repeat(pad - left);
      }
      return content + " ".repeat(pad);
    });
  }
  /**
   * Returns the shared gradient when every series resolves to the same
   * gradient (the chart-wide `color: 'gradient(...)'` case). Mixed colors
   * or per-series-different gradients return undefined.
   */
  sharedSeriesGradient(seriesColors) {
    if (seriesColors.length === 0) return void 0;
    const themeColors = this.getThemeColors();
    const normalize = (c) => typeof c === "string" ? normalizeColor(c) : c;
    const first = normalize(seriesColors[0]);
    if (!isGradientObject(first)) return void 0;
    const firstGrad = first;
    const signature = (g) => g.colors.map((c) => parseColorToRgb(c, themeColors).join(",")).join("|");
    const ref = signature(firstGrad);
    for (let i = 1; i < seriesColors.length; i++) {
      const norm = normalize(seriesColors[i]);
      if (!isGradientObject(norm)) return void 0;
      if (signature(norm) !== ref) return void 0;
    }
    return firstGrad;
  }
  /**
   * Render legend labels as a single combined block painted with the chart's
   * shared gradient. Labels are joined by single spaces and the gradient
   * stretches across the full combined width; rows wrap to `chartWidth`
   * while the gradient continues smoothly across wraps.
   */
  buildGradientLegendRows(labels, gradient, config, chartWidth) {
    const combined = labels.join(" ");
    const total = combined.length;
    if (total === 0) return [];
    const isSeparator = new Array(total).fill(false);
    let cursor = 0;
    for (let i = 0; i < labels.length; i++) {
      cursor += labels[i].length;
      if (i < labels.length - 1) {
        isSeparator[cursor] = true;
        cursor += 1;
      }
    }
    const cells = [];
    for (let i = 0; i < total; i++) {
      if (isSeparator[i]) {
        cells.push(" ");
        continue;
      }
      const position = total > 1 ? i / (total - 1) : 0;
      const [br, bg, bb] = this.getColorAtPosition(gradient, position);
      const [fr, fg, fb] = this.contrastFgRgb([br, bg, bb]);
      cells.push(
        `\x1B[48;2;${br};${bg};${bb}m\x1B[38;2;${fr};${fg};${fb}m${combined[i]}\x1B[39m\x1B[49m`
      );
    }
    const rows = [];
    for (let start = 0; start < total; start += chartWidth) {
      const end = Math.min(start + chartWidth, total);
      const content = cells.slice(start, end).join("");
      const visualWidth = end - start;
      const pad = Math.max(0, chartWidth - visualWidth);
      if (config.align === "right") rows.push(" ".repeat(pad) + content);
      else if (config.align === "center") {
        const left = Math.floor(pad / 2);
        rows.push(" ".repeat(left) + content + " ".repeat(pad - left));
      } else rows.push(content + " ".repeat(pad));
    }
    return rows;
  }
  resolveLegendSwatchRgb(color) {
    if (!color) return void 0;
    const normalized = typeof color === "string" ? normalizeColor(color) : color;
    if (isGradientObject(normalized)) {
      const grad = normalized;
      if (grad.colors.length === 0) return void 0;
      return parseColorToRgb(grad.colors[0], this.getThemeColors());
    }
    return parseColorToRgb(normalized, this.getThemeColors());
  }
  contrastFgRgb(rgb) {
    const luminance = 0.299 * rgb[0] + 0.587 * rgb[1] + 0.114 * rgb[2];
    return luminance > 128 ? [0, 0, 0] : [255, 255, 255];
  }
};
var line_default = LineChartFormatter;

// formatters/step.ts
var STEP_SHARP = {
  horizontal: "\u2500",
  vertical: "\u2502",
  cornerTopLeft: "\u250C",
  cornerTopRight: "\u2510",
  cornerBottomLeft: "\u2514",
  cornerBottomRight: "\u2518"
};
var STEP_SMOOTH = {
  horizontal: "\u2500",
  vertical: "\u2502",
  cornerTopLeft: "\u256D",
  cornerTopRight: "\u256E",
  cornerBottomLeft: "\u2570",
  cornerBottomRight: "\u256F"
};
var StepChartFormatter = class extends line_default {
  /**
   * Step charts have no diagonal-spacing constraint, so the natural step
   * width is `requestedWidth / (n-1)`. We only widen the chart to ensure
   * adjacent x-axis labels keep ≥ 1 space.
   */
  prepareLayout(points, requestedWidth) {
    if (points.length === 0) return requestedWidth;
    if (points.length === 1) {
      points[0].col = Math.floor(requestedWidth / 2);
      return requestedWidth;
    }
    const minDiffs = [0];
    for (let i = 1; i < points.length; i++) {
      const L1 = this.labelLength(points[i - 1].point?.label);
      const L2 = this.labelLength(points[i].point?.label);
      minDiffs.push(this.minLabelColDiff(L1, L2));
    }
    const evenStep = (requestedWidth - 1) / (points.length - 1);
    points[0].col = 0;
    for (let i = 1; i < points.length; i++) {
      const evenCol = Math.round(i * evenStep);
      const minCol = points[i - 1].col + minDiffs[i];
      points[i].col = Math.max(evenCol, minCol);
    }
    const firstLabel = points[0].point?.label ?? "";
    const lastLabel = points[points.length - 1].point?.label ?? "";
    const leadPad = Math.floor(this.labelLength(firstLabel) / 2);
    const trailPad = Math.max(0, this.labelLength(lastLabel) - 1 - Math.floor(this.labelLength(lastLabel) / 2));
    if (leadPad > 0) {
      for (const p of points) p.col += leadPad;
    }
    return Math.max(requestedWidth, points[points.length - 1].col + 1 + trailPad);
  }
  prepareLayoutMulti(seriesPoints, requestedWidth) {
    if (seriesPoints.length === 0) return requestedWidth;
    const firstSeries = seriesPoints[0];
    const width = this.prepareLayout(firstSeries, requestedWidth);
    for (let s = 1; s < seriesPoints.length; s++) {
      const points = seriesPoints[s];
      for (let i = 0; i < points.length && i < firstSeries.length; i++) {
        points[i].col = firstSeries[i].col;
      }
    }
    return width;
  }
  minLabelColDiff(L1, L2) {
    if (L1 <= 0 && L2 <= 0) return 1;
    return Math.max(1, Math.ceil(L1 / 2) + Math.floor(L2 / 2) + 1);
  }
  drawLine(grid, points, chartWidth) {
    return this.drawStep(grid, void 0, points, chartWidth, void 0);
  }
  drawLineColored(grid, gridColors, points, chartWidth, color) {
    return this.drawStep(grid, gridColors, points, chartWidth, color);
  }
  drawStep(grid, gridColors, points, chartWidth, color) {
    const chars = this.options.variant === "sharp" ? STEP_SHARP : STEP_SMOOTH;
    const lineBottom = new Array(chartWidth).fill(-1);
    if (points.length === 0) return lineBottom;
    const seriesStart = points[0].col;
    const seriesEnd = points[points.length - 1].col;
    const height = grid.length;
    const put = (row, col, ch) => {
      if (row >= 0 && row < grid.length && col >= 0 && col < chartWidth) {
        grid[row][col] = ch;
        if (gridColors) {
          const c = this.resolveCellColor(color, col, row, seriesStart, seriesEnd, height);
          if (c) gridColors[row][col] = c;
        }
        lineBottom[col] = row;
      }
    };
    const first = points[0];
    const last = points[points.length - 1];
    for (let col = 0; col <= first.col; col++) {
      put(first.row, col, chars.horizontal);
    }
    for (let i = 0; i < points.length - 1; i++) {
      const p1 = points[i];
      const p2 = points[i + 1];
      if (p2.col <= p1.col) continue;
      if (p1.row === p2.row) {
        for (let col = p1.col + 1; col <= p2.col; col++) {
          put(p1.row, col, chars.horizontal);
        }
        continue;
      }
      for (let col = p1.col + 1; col < p2.col; col++) {
        put(p1.row, col, chars.horizontal);
      }
      const goingUp = p2.row < p1.row;
      const topRow = Math.min(p1.row, p2.row);
      const bottomRow = Math.max(p1.row, p2.row);
      if (goingUp) {
        put(bottomRow, p2.col, chars.cornerBottomRight);
        put(topRow, p2.col, chars.cornerTopLeft);
      } else {
        put(topRow, p2.col, chars.cornerTopRight);
        put(bottomRow, p2.col, chars.cornerBottomLeft);
      }
      for (let row = topRow + 1; row < bottomRow; row++) {
        put(row, p2.col, chars.vertical);
      }
      lineBottom[p2.col] = bottomRow;
    }
    for (let col = last.col + 1; col < chartWidth; col++) {
      put(last.row, col, chars.horizontal);
    }
    return lineBottom;
  }
};
var step_default = StepChartFormatter;

// formatters/scatter.ts
var ScatterChartFormatter = class extends line_default {
  constructor(options) {
    super(options);
  }
  shouldDrawPoints() {
    return true;
  }
  /**
   * Scatter has no diagonals to align — points map directly to evenly
   * spaced columns from `mapPointsToGrid`. We still enforce ≥1 space
   * between adjacent x-axis labels by widening the chart only as much
   * as the label-spacing constraint requires.
   */
  prepareLayout(points, requestedWidth) {
    if (points.length === 0) return requestedWidth;
    if (points.length === 1) return requestedWidth;
    const minDiffs = [0];
    for (let i = 1; i < points.length; i++) {
      const L1 = this.labelLength(points[i - 1].point?.label);
      const L2 = this.labelLength(points[i].point?.label);
      minDiffs.push(this.minLabelColDiff(L1, L2));
    }
    const evenStep = (requestedWidth - 1) / (points.length - 1);
    let prevCol = 0;
    points[0].col = 0;
    for (let i = 1; i < points.length; i++) {
      const evenCol = Math.round(i * evenStep);
      const minCol = prevCol + minDiffs[i];
      const col = Math.max(evenCol, minCol);
      points[i].col = col;
      prevCol = col;
    }
    const firstLabel = points[0].point?.label ?? "";
    const lastLabel = points[points.length - 1].point?.label ?? "";
    const leadPad = Math.floor(this.labelLength(firstLabel) / 2);
    const trailPad = Math.max(0, this.labelLength(lastLabel) - 1 - Math.floor(this.labelLength(lastLabel) / 2));
    if (leadPad > 0) {
      for (const p of points) p.col += leadPad;
    }
    return Math.max(requestedWidth, points[points.length - 1].col + 1 + trailPad);
  }
  prepareLayoutMulti(seriesPoints, requestedWidth) {
    if (seriesPoints.length === 0) return requestedWidth;
    const firstSeries = seriesPoints[0];
    const width = this.prepareLayout(firstSeries, requestedWidth);
    for (let s = 1; s < seriesPoints.length; s++) {
      const points = seriesPoints[s];
      for (let i = 0; i < points.length && i < firstSeries.length; i++) {
        points[i].col = firstSeries[i].col;
      }
    }
    return width;
  }
  minLabelColDiff(L1, L2) {
    if (L1 <= 0 && L2 <= 0) return 1;
    return Math.max(1, Math.ceil(L1 / 2) + Math.floor(L2 / 2) + 1);
  }
  drawLine(_grid, _points, chartWidth) {
    return new Array(chartWidth).fill(-1);
  }
  drawLineColored(_grid, _gridColors, _points, chartWidth) {
    return new Array(chartWidth).fill(-1);
  }
  /**
   * Single-series scatter always takes the colored path so per-point colors
   * (from `color: 'auto'` cycling the palette, or per-point `{value, color}`)
   * land in `gridColors`. Unlike line/step, scatter has no series-wide line
   * color to fall back on — markers are independent.
   */
  format(chartData) {
    const chart = [...chartData.values()];
    if (chart.length === 0) return "";
    const height = this.options.height || 10;
    const width = this.options.width || 80;
    const maxValue = this.options.max.value || 1;
    const minValue = this.options.max.min ?? 0;
    const yAxisTicks = this.buildYAxisTicks(minValue, maxValue, height);
    const yLabelWidth = Math.max(...yAxisTicks.map((t) => t.label.length));
    const requestedWidth = width - yLabelWidth - 1;
    if (requestedWidth < 2) return "";
    const points = this.mapPointsToGrid(chart, requestedWidth, height, minValue, maxValue);
    const chartWidth = this.prepareLayout(points, requestedWidth);
    const grid = this.buildGrid(chartWidth, height);
    const gridColors = Array.from(
      { length: height },
      () => new Array(chartWidth).fill(void 0)
    );
    const seriesColor = this.options._seriesColors?.[0];
    const pointChar = this.options.pointChar || "\u25C8";
    this.drawPointsColored(grid, gridColors, points, pointChar, seriesColor);
    this.applyAnimationClip(grid, gridColors, chartWidth);
    return this.compose(grid, yAxisTicks, yLabelWidth, chartWidth, height, points, gridColors);
  }
  /**
   * Override: prefer the per-point color (set by the processor's auto-color
   * pass or the user) over the series-level fallback, so single-series
   * scatter with `color: 'auto'` actually cycles the palette per marker.
   */
  drawPointsColored(grid, gridColors, points, char, color) {
    if (points.length === 0) return;
    const seriesStart = points[0].col;
    const seriesEnd = points[points.length - 1].col;
    const height = grid.length;
    for (const { col, row, point } of points) {
      if (row < 0 || row >= grid.length || col < 0 || col >= grid[0].length) continue;
      grid[row][col] = char;
      const effective = point.color || color;
      if (effective === void 0 || effective === "") continue;
      const cellColor = this.resolveCellColor(effective, col, row, seriesStart, seriesEnd, height);
      if (cellColor) gridColors[row][col] = cellColor;
    }
  }
  /**
   * Color each x-axis label with its point's color when `colorLabels` is on.
   * Labels are centered on each point's column; we render label-by-label so
   * neighboring labels keep distinct colors instead of bleeding together.
   */
  formatXLabels(points, offset, chartWidth) {
    if (!this.options.colorLabels) {
      return super.formatXLabels(points, offset, chartWidth);
    }
    const seriesColors = this.options._seriesColors ?? [];
    const seriesColor = seriesColors[0];
    const normalizedSeries = typeof seriesColor === "string" ? normalizeColor(seriesColor) : seriesColor;
    const chartGradient = this.sharedSeriesGradient(seriesColors) ?? (isGradientObject(normalizedSeries) ? normalizedSeries : void 0);
    const seriesStart = points[0]?.col ?? 0;
    const seriesEnd = points[points.length - 1]?.col ?? 0;
    const span = seriesEnd - seriesStart;
    const colorForPoint = (col, pointColor) => {
      if (chartGradient) {
        const position = span > 0 ? (col - seriesStart) / span : 0;
        const [r, g, b] = this.getColorAtPosition(chartGradient, position);
        return `#${r.toString(16).padStart(2, "0")}${g.toString(16).padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
      }
      if (pointColor) return pointColor;
      return typeof normalizedSeries === "string" ? normalizedSeries : void 0;
    };
    const cells = Array.from({ length: chartWidth }, () => ({ ch: " " }));
    for (const { col, point } of points) {
      const label = point.label;
      const start = Math.max(0, col - Math.floor(label.length / 2));
      const labelColor = colorForPoint(col, point.color);
      for (let i = 0; i < label.length && start + i < chartWidth; i++) {
        cells[start + i] = { ch: label[i], color: labelColor };
      }
    }
    let out = " ".repeat(offset);
    for (const { ch, color } of cells) {
      out += color ? this.colorify(ch, color) : ch;
    }
    return out;
  }
};
var scatter_default = ScatterChartFormatter;

// formatters/candlestick.ts
var WICK_CHAR = "\u2502";
var BODY_CHAR = "\u2588";
var DOJI_CHAR = "\u2500";
var CandlestickChartFormatter = class extends line_default {
  constructor(options) {
    super(options);
  }
  /**
   * Stub — candlestick uses its own `layout()` for column placement
   * (auto-fill or explicit padding). The base `calculateBarDimensions`
   * machinery is bypassed for this chart type.
   */
  calculateDefaultDimensions(_barCount, _totalSize, _charWidth) {
    return { barSize: this.options.barSize ?? 1, padding: this.options.padding ?? 0 };
  }
  format(chartData) {
    const chart = [...chartData.values()];
    if (chart.length === 0) return "";
    const height = this.options.height || 10;
    const width = this.options.width || 80;
    const maxValue = this.options.max.value || 1;
    const minValue = this.options.max.min ?? 0;
    const yAxisTicks = this.buildYAxisTicks(minValue, maxValue, height);
    const yLabelWidth = Math.max(...yAxisTicks.map((t) => t.label.length));
    const requestedWidth = Math.max(2, width - yLabelWidth - 1);
    const { barSize, candleStarts, chartWidth } = this.layout(chart, requestedWidth);
    const grid = this.buildGrid(chartWidth, height);
    const gridColors = Array.from(
      { length: height },
      () => new Array(chartWidth).fill(void 0)
    );
    const points = [];
    for (let i = 0; i < chart.length; i++) {
      const point = chart[i];
      const slotStart = candleStarts[i];
      const candleCenter = slotStart + Math.floor(barSize / 2);
      this.drawCandle(grid, gridColors, point, i, chart.length, slotStart, barSize, height, minValue, maxValue);
      points.push({ col: candleCenter, row: 0, point });
    }
    this.applyAnimationClip(grid, gridColors, chartWidth);
    this.prepareCandlestickLegend();
    return this.compose(grid, yAxisTicks, yLabelWidth, chartWidth, height, points, gridColors);
  }
  /**
   * Determine candle column positions.
   *
   * Two modes:
   *   • **Explicit padding** — `barSize` body cols + `padding` gap cols
   *     per candle, no width-fill. Honors what the user asked for.
   *   • **Auto-fill** (default) — distribute candles evenly across the
   *     full available width. The first candle sits at col `leadPad`,
   *     the last at `chartWidth - barSize - trailPad`; the in-between
   *     are rounded to the nearest col so spacing alternates by ±1.
   *     `leadPad` / `trailPad` reserve room for the first / last
   *     x-axis label to render without clipping.
   */
  layout(chart, requestedWidth) {
    const barCount = chart.length;
    const barSize = this.options.barSize ?? 1;
    if (this.options.padding !== void 0) {
      const padding = this.options.padding;
      const slotWidth = barSize + padding;
      const starts2 = Array.from({ length: barCount }, (_, i) => i * slotWidth);
      const totalDataWidth = barCount > 0 ? slotWidth * barCount - padding : 0;
      return {
        barSize,
        candleStarts: starts2,
        chartWidth: Math.max(totalDataWidth, barSize)
      };
    }
    if (barCount === 0) return { barSize, candleStarts: [], chartWidth: requestedWidth };
    const firstLabel = chart[0]?.label ?? "";
    const lastLabel = chart[barCount - 1]?.label ?? "";
    const leadPad = Math.floor(firstLabel.length / 2);
    const trailPad = Math.max(0, lastLabel.length - 1 - Math.floor(lastLabel.length / 2));
    if (barCount === 1) {
      return {
        barSize,
        candleStarts: [leadPad],
        chartWidth: Math.max(requestedWidth, leadPad + barSize + trailPad)
      };
    }
    const usable = Math.max(barSize, requestedWidth - leadPad - trailPad);
    const step = (usable - barSize) / (barCount - 1);
    const starts = Array.from({ length: barCount }, (_, i) => leadPad + Math.round(i * step));
    const lastStart = starts[barCount - 1];
    const chartWidth = Math.max(requestedWidth, lastStart + barSize + trailPad);
    return { barSize, candleStarts: starts, chartWidth };
  }
  /**
   * Candlestick is single-series. If a caller still routes here, just
   * format the first chart and ignore the rest.
   */
  formatMulti(charts) {
    return this.format(charts[0] ?? /* @__PURE__ */ new Map());
  }
  /**
   * Color each x-axis label with its candle's resolved color so labels
   * track bull/bear, per-candle overrides, and time-decay gradients the
   * same way the bodies do. Falls back to the uncolored base when
   * `colorLabels` is off.
   */
  formatXLabels(points, offset, chartWidth) {
    if (!this.options.colorLabels) {
      return super.formatXLabels(points, offset, chartWidth);
    }
    const cells = Array.from({ length: chartWidth }, () => ({ ch: " " }));
    const total = points.length;
    for (let i = 0; i < points.length; i++) {
      const { col, point } = points[i];
      const label = point.label ?? "";
      if (!label) continue;
      let labelColor;
      if (point.ohlc) {
        const [open, , , close] = point.ohlc;
        labelColor = this.resolveCandleColor(point, close >= open, i, total);
      }
      const start = Math.max(0, col - Math.floor(label.length / 2));
      for (let j = 0; j < label.length && start + j < chartWidth; j++) {
        cells[start + j] = { ch: label[j], color: labelColor };
      }
    }
    let out = " ".repeat(offset);
    for (const { ch, color } of cells) {
      if (!color || ch === " ") {
        out += ch;
      } else if (color.startsWith("\x1B[38;2;")) {
        out += `${color}${ch}\x1B[39m`;
      } else {
        out += this.colorify(ch, color);
      }
    }
    return out;
  }
  drawCandle(grid, gridColors, point, index, total, slotStart, barSize, height, minValue, maxValue) {
    if (!point.ohlc) return;
    const [open, high, low, close] = point.ohlc;
    const range = maxValue - minValue || 1;
    const valueToRow = (v) => {
      const normalized = (v - minValue) / range;
      const row = height - 1 - Math.round(normalized * (height - 1));
      return Math.max(0, Math.min(height - 1, row));
    };
    const lowRow = valueToRow(low);
    const highRow = valueToRow(high);
    const openRow = valueToRow(open);
    const closeRow = valueToRow(close);
    const isBull = close >= open;
    const cellColor = this.resolveCandleColor(point, isBull, index, total);
    const wickCol = slotStart + Math.floor(barSize / 2);
    const gridWidth = grid[0]?.length ?? 0;
    if (wickCol >= 0 && wickCol < gridWidth) {
      for (let row = highRow; row <= lowRow; row++) {
        grid[row][wickCol] = WICK_CHAR;
        if (cellColor) gridColors[row][wickCol] = cellColor;
      }
    }
    const bodyTop = Math.min(openRow, closeRow);
    const bodyBottom = Math.max(openRow, closeRow);
    const isDoji = openRow === closeRow;
    for (let col = slotStart; col < slotStart + barSize && col < gridWidth; col++) {
      for (let row = bodyTop; row <= bodyBottom; row++) {
        grid[row][col] = isDoji ? DOJI_CHAR : BODY_CHAR;
        if (cellColor) gridColors[row][col] = cellColor;
      }
    }
  }
  /**
   * Resolve the body/wick color for one candle to a value the cell
   * colorizer can apply directly. Order of precedence:
   *   1. Per-candle `point.color` override (wins regardless of bull/bear)
   *   2. `_bullColor` / `_bearColor` based on close ≥ open
   *
   * Gradient values (in any of these slots) are pre-interpolated to a
   * single RGB ANSI escape per candle index (time-decay coloring) so the
   * candle paints uniformly rather than stretching the gradient inside
   * each candle.
   */
  resolveCandleColor(point, isBull, index, total) {
    const override = point.color;
    if (override) return this.flattenColor(override, index, total);
    const base = isBull ? this.options._bullColor : this.options._bearColor;
    return this.flattenColor(base, index, total);
  }
  flattenColor(color, index, total) {
    if (color === void 0 || color === "") return void 0;
    const normalized = typeof color === "string" ? normalizeColor(color) : color;
    if (!isGradientObject(normalized)) return normalized;
    const grad = normalized;
    const position = total > 1 ? index / (total - 1) : 0;
    const finalPosition = grad.reverse ? 1 - position : position;
    const [r, g, b] = this.getColorAtPosition(grad, finalPosition);
    return `\x1B[38;2;${r};${g};${b}m`;
  }
  /**
   * Hand the inherited legend builder a synthetic two-series setup
   * (`_seriesColors = [bull, bear]`) when the legend is enabled and the
   * bull/bear colors actually differ. Default labels are `Bullish` /
   * `Bearish`; a user-supplied `legend.values` array overrides them.
   *
   * When bull and bear are the same color, a two-entry legend would be
   * meaningless — leave `_seriesColors` empty so the inherited
   * `resolveLegendConfig` skips it.
   */
  prepareCandlestickLegend() {
    if (!this.options.legend) return;
    const bull = this.options._bullColor;
    const bear = this.options._bearColor;
    if (bull === void 0 && bear === void 0) return;
    const sameColor = this.colorsEqual(bull, bear);
    if (sameColor) return;
    this.options._seriesColors = [bull, bear];
    const legend = typeof this.options.legend === "object" ? this.options.legend : {};
    if (!legend.values || legend.values.length === 0) {
      this.options.legend = { ...legend, values: ["Bullish", "Bearish"] };
    }
  }
  colorsEqual(a, b) {
    if (a === b) return true;
    if (typeof a === "string" && typeof b === "string") {
      return normalizeColor(a) === normalizeColor(b);
    }
    if (typeof a !== "string" && typeof b !== "string" && a && b) {
      const themeColors = this.getThemeColors();
      const sig = (g) => g.colors.map((c) => parseColorToRgb(c, themeColors).join(",")).join("|");
      return sig(a) === sig(b);
    }
    return false;
  }
};
var candlestick_default = CandlestickChartFormatter;

// formatters/status.ts
var BLOCK_CHAR = "\u2588";
var StatusChartFormatter = class extends line_default {
  constructor(options) {
    super(options);
  }
  /**
   * Stub — status uses its own grid layout, not the bar dimension model.
   */
  calculateDefaultDimensions(_barCount, _totalSize, _charWidth) {
    return { barSize: this.options.barSize ?? 2, padding: this.options.padding ?? 1 };
  }
  format(chartData) {
    const chart = [...chartData.values()];
    if (chart.length === 0) return "";
    if (chart.some((p) => p.statuses)) {
      return this.formatRows(chart);
    }
    const barSize = this.options.barSize ?? 2;
    const cellHeight = Math.max(1, Math.ceil(barSize / 2));
    const padding = this.options.padding ?? 1;
    const total = chart.length;
    const cols = this.resolveCols(total, barSize, padding);
    const rows = Math.ceil(total / cols);
    const showLabels = this.options.labels !== false;
    const labelRows = showLabels ? 1 : 0;
    const cellRowHeight = cellHeight + labelRows;
    const chartWidth = cols * barSize + Math.max(0, cols - 1) * padding;
    const chartHeight = rows * cellRowHeight + Math.max(0, rows - 1) * padding;
    const grid = this.buildGrid(chartWidth, chartHeight);
    const gridColors = Array.from(
      { length: chartHeight },
      () => new Array(chartWidth).fill(void 0)
    );
    const visibleCount = this.computeVisibleCount(total);
    for (let i = 0; i < total; i++) {
      const point = chart[i];
      const gridRow = Math.floor(i / cols);
      const gridCol = i % cols;
      const xStart = gridCol * (barSize + padding);
      const yStart = gridRow * (cellRowHeight + padding);
      if (i >= visibleCount) continue;
      const cellColor = this.resolveStatusColor(point, i, total);
      for (let dy = 0; dy < cellHeight; dy++) {
        for (let dx = 0; dx < barSize; dx++) {
          const col = xStart + dx;
          const row = yStart + dy;
          if (col < chartWidth && row < chartHeight) {
            grid[row][col] = BLOCK_CHAR;
            if (cellColor) gridColors[row][col] = cellColor;
          }
        }
      }
      if (showLabels && point.label) {
        const label = point.label;
        const labelRow = yStart + cellHeight;
        const cellCenter = xStart + Math.floor(barSize / 2);
        const labelStart = Math.max(xStart, cellCenter - Math.floor(label.length / 2));
        const cellEnd = xStart + barSize;
        for (let j = 0; j < label.length; j++) {
          const col = labelStart + j;
          if (col >= cellEnd || col >= chartWidth) break;
          grid[labelRow][col] = label[j];
          if (this.options.colorLabels !== false && cellColor) {
            gridColors[labelRow][col] = cellColor;
          }
        }
      }
    }
    this.prepareStatusLegend();
    return this.composeStatus(grid, gridColors, chartWidth, chartHeight);
  }
  /**
   * Row mode: render one row per point, with the label on the left (y-axis)
   * and an array of colored cells extending across the x-axis. Scalar points
   * become a single-cell row; ragged arrays are padded with blank space on
   * the right (no cell drawn, no `padding` character either — keeps the
   * x-axis aligned across rows).
   */
  formatRows(chart) {
    const barSize = this.options.barSize ?? 2;
    const cellHeight = Math.max(1, Math.ceil(barSize / 2));
    const padding = this.options.padding ?? 1;
    const showLabels = this.options.labels !== false;
    const cellsPerRow = chart.map((p) => p.statuses ? p.statuses.length : p.status !== void 0 ? 1 : 0);
    const maxCells = Math.max(0, ...cellsPerRow);
    const labelWidth = showLabels ? Math.max(0, ...chart.map((p) => p.label ? p.label.length : 0)) : 0;
    const labelGap = labelWidth > 0 ? 1 : 0;
    const cellsAreaWidth = maxCells * barSize + Math.max(0, maxCells - 1) * padding;
    const chartWidth = labelWidth + labelGap + cellsAreaWidth;
    const rows = chart.length;
    const chartHeight = rows * cellHeight + Math.max(0, rows - 1) * padding;
    const grid = this.buildGrid(chartWidth, chartHeight);
    const gridColors = Array.from(
      { length: chartHeight },
      () => new Array(chartWidth).fill(void 0)
    );
    const totalCells = cellsPerRow.reduce((sum, n) => sum + n, 0);
    const visibleCount = this.computeVisibleCount(totalCells);
    let cellIndex = 0;
    for (let r = 0; r < rows; r++) {
      const point = chart[r];
      const yStart = r * (cellHeight + padding);
      const cells = point.statuses ? point.statuses : point.status !== void 0 ? [point.status] : [];
      if (showLabels && point.label && labelWidth > 0) {
        const labelRow = yStart + Math.floor(cellHeight / 2);
        const padded = point.label.padStart(labelWidth);
        for (let i = 0; i < padded.length; i++) {
          if (i < chartWidth) grid[labelRow][i] = padded[i];
        }
      }
      for (let c = 0; c < cells.length; c++) {
        const xStart = labelWidth + labelGap + c * (barSize + padding);
        const cellPoint = { ...point, status: cells[c] };
        const cellColor = this.resolveStatusColor(cellPoint, cellIndex, totalCells);
        if (cellIndex < visibleCount) {
          for (let dy = 0; dy < cellHeight; dy++) {
            for (let dx = 0; dx < barSize; dx++) {
              const col = xStart + dx;
              const row = yStart + dy;
              if (col < chartWidth && row < chartHeight) {
                grid[row][col] = BLOCK_CHAR;
                if (cellColor) gridColors[row][col] = cellColor;
              }
            }
          }
        }
        cellIndex++;
      }
    }
    this.prepareStatusLegend();
    return this.composeStatus(grid, gridColors, chartWidth, chartHeight);
  }
  /**
   * Status is single-series. If a caller routes here, fall back to the
   * single-chart format with the first dataset.
   */
  formatMulti(charts) {
    return this.format(charts[0] ?? /* @__PURE__ */ new Map());
  }
  /**
   * Pick the column count for the grid. With `width` set, pack as many
   * `barSize`-wide cells as fit (separated by `padding`). Otherwise fall
   * back to a roughly-square grid via `ceil(sqrt(N))`. Always clamped to
   * `[1, total]` — no empty columns, no more columns than items.
   */
  resolveCols(total, barSize, padding) {
    const width = this.options.width;
    if (typeof width === "number" && width > 0) {
      const fit = Math.floor((width + padding) / (barSize + padding));
      return Math.max(1, Math.min(total, fit));
    }
    return Math.max(1, Math.min(total, Math.ceil(Math.sqrt(total))));
  }
  /**
   * Animation reveal works in scan order: at progress p, the first
   * `floor(p * total)` cells are drawn; the rest are blank.
   */
  computeVisibleCount(total) {
    const progress = this.options._animationProgress;
    if (progress === void 0 || progress >= 1) return total;
    if (progress <= 0) return 0;
    return Math.max(0, Math.min(total, Math.floor(progress * total)));
  }
  /**
   * Custom compose for status — no y-axis or bottom rule. Stitches
   * title + legend(top) + grid rows + legend(bottom) in the same order
   * the line formatter does, just without the chart-frame chrome.
   */
  composeStatus(grid, gridColors, chartWidth, chartHeight) {
    const lines = [];
    const title = this.formatTitle(chartWidth);
    if (title) lines.push(title);
    const legendConfig = this.resolveLegendConfig();
    const legendRows = legendConfig ? this.buildLegendRows(legendConfig, chartWidth) : [];
    if (legendConfig?.position === "top") {
      for (const row of legendRows) lines.push(row);
      lines.push(" ".repeat(chartWidth));
    }
    for (let row = 0; row < chartHeight; row++) {
      lines.push(this.colorizeRow(grid[row], row, chartHeight, chartWidth, gridColors[row]));
    }
    if (legendConfig?.position === "bottom") {
      lines.push(" ".repeat(chartWidth));
      for (const row of legendRows) lines.push(row);
    }
    return lines.join("\n");
  }
  /**
   * Resolve the color for a single cell. Order of precedence:
   *   1. Per-point `point.color` override
   *   2. The status's color from the `_statusColors` map
   *
   * Gradients flatten to a single RGB ANSI escape per cell, interpolated
   * across the global cell index (0..total-1) — same pattern candlestick
   * uses for time-decay coloring.
   */
  resolveStatusColor(point, index, total) {
    if (point.color) return this.flattenStatusColor(point.color, index, total);
    const map = this.options._statusColors;
    if (!map || !point.status) return void 0;
    return this.flattenStatusColor(map[point.status], index, total);
  }
  flattenStatusColor(color, index, total) {
    if (color === void 0 || color === "") return void 0;
    const normalized = typeof color === "string" ? normalizeColor(color) : color;
    if (!isGradientObject(normalized)) return normalized;
    const grad = normalized;
    const position = total > 1 ? index / (total - 1) : 0;
    const finalPosition = grad.reverse ? 1 - position : position;
    const [r, g, b] = this.getColorAtPosition(grad, finalPosition);
    return `\x1B[38;2;${r};${g};${b}m`;
  }
  /**
   * Hand the inherited legend builder one entry per status key. Each
   * legend swatch shows the color from the status map. Default values
   * are the status keys themselves; a user-supplied `legend.values`
   * array overrides them.
   */
  prepareStatusLegend() {
    if (!this.options.legend) return;
    const map = this.options._statusColors;
    if (!map) return;
    const statuses = Object.keys(map);
    if (statuses.length === 0) return;
    this.options._seriesColors = statuses.map((s) => map[s]);
    const legend = typeof this.options.legend === "object" ? this.options.legend : {};
    if (!legend.values || legend.values.length === 0) {
      this.options.legend = { ...legend, values: statuses };
    }
  }
};
var status_default = StatusChartFormatter;

// chartscii.ts
var SERIES_AUTO_COLORS = ["red", "green", "yellow", "blue", "purple", "cyan", "pink", "orange", "marine"];
var DEFAULT_BULL_COLOR = "green";
var DEFAULT_BEAR_COLOR = "red";
function isPointChartType(type) {
  return type === "line" || type === "step" || type === "scatter";
}
function supportsArrayColor(type) {
  return isPointChartType(type) || type === "candlestick";
}
function usesProgressReveal(type) {
  return isPointChartType(type) || type === "candlestick" || type === "status";
}
function isStatusColorMap(color) {
  return color !== null && typeof color === "object" && !Array.isArray(color) && color.type !== "gradient";
}
function normalizeStatusInput(data) {
  return data.map((item) => {
    if (typeof item === "number" || typeof item === "string") {
      return { value: item };
    }
    if (item && typeof item === "object" && item.status !== void 0 && item.value === void 0) {
      const { status, label, color } = item;
      return { value: status, label, color };
    }
    return item;
  });
}
function resolveStatusColors(color) {
  if (isStatusColorMap(color)) return color;
  return void 0;
}
function normalizeCandlestickInput(data) {
  return data.map((item) => {
    if (Array.isArray(item)) {
      return { value: item.slice(0, 4) };
    }
    return item;
  });
}
function resolveCandlestickColors(color) {
  if (color === void 0 || color === "auto" || isStatusColorMap(color)) {
    return { bullColor: DEFAULT_BULL_COLOR, bearColor: DEFAULT_BEAR_COLOR };
  }
  if (Array.isArray(color)) {
    return { bullColor: color[0], bearColor: color[1] ?? color[0] };
  }
  return { bullColor: color, bearColor: color };
}
function resolveSeriesColors(color, seriesCount) {
  if (color === "auto") {
    return Array.from({ length: seriesCount }, (_, i) => SERIES_AUTO_COLORS[i % SERIES_AUTO_COLORS.length]);
  }
  if (Array.isArray(color)) {
    return Array.from({ length: seriesCount }, (_, i) => color[i]);
  }
  if (isStatusColorMap(color)) {
    return Array.from({ length: seriesCount }, () => void 0);
  }
  return Array.from({ length: seriesCount }, () => color);
}
function narrowArrayColor(color) {
  if (Array.isArray(color)) return color[0];
  if (isStatusColorMap(color)) return void 0;
  return color;
}
function isMultiSeriesData(data) {
  return Array.isArray(data) && data.length > 0 && Array.isArray(data[0]);
}
var easings = {
  linear: (t) => t,
  easeIn: (t) => t * t,
  easeOut: (t) => t * (2 - t),
  easeInOut: (t) => t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t
};
function roundValue(value, decimals) {
  const factor = Math.pow(10, decimals);
  return Math.round(value * factor) / factor;
}
function getOriginalLabel(item) {
  if (typeof item === "number") {
    return item.toString();
  }
  if (item.label) {
    return item.label;
  }
  if (Array.isArray(item.value)) {
    let total = 0;
    for (const v of item.value) {
      total += typeof v === "number" ? v : v.value;
    }
    return total.toString();
  }
  return item.value.toString();
}
function scaleInputData(data, progress, decimals = 2, preserveLabels = false) {
  return data.map((item) => {
    const originalLabel = preserveLabels ? getOriginalLabel(item) : void 0;
    if (typeof item === "number") {
      const scaledValue = roundValue(item * progress, decimals);
      return preserveLabels ? { value: scaledValue, label: originalLabel } : scaledValue;
    }
    if (Array.isArray(item.value)) {
      return {
        ...item,
        ...originalLabel && { label: originalLabel },
        value: item.value.map(
          (v) => typeof v === "number" ? roundValue(v * progress, decimals) : { ...v, value: roundValue(v.value * progress, decimals) }
        )
      };
    }
    return {
      ...item,
      ...originalLabel && { label: originalLabel },
      value: roundValue(item.value * progress, decimals)
    };
  });
}
function getMaxValue(data) {
  let maxVal = 0;
  for (const item of data) {
    let value;
    if (typeof item === "number") {
      value = item;
    } else if (Array.isArray(item.value)) {
      value = 0;
      for (const v of item.value) {
        value += typeof v === "number" ? v : v.value;
      }
    } else {
      value = item.value;
    }
    if (value > maxVal) maxVal = value;
  }
  return maxVal;
}
var Chartscii = class _Chartscii {
  constructor(data, options) {
    const config = createOptions(options || {});
    this.options = options;
    this.originalRawData = data;
    const chartType = config.type || "bar";
    if (!supportsArrayColor(chartType) && Array.isArray(config.color)) {
      config.color = narrowArrayColor(config.color);
    }
    if (chartType !== "status" && isStatusColorMap(config.color)) {
      config.color = void 0;
    }
    if (chartType === "status") {
      const statusData = normalizeStatusInput(data);
      this.originalData = statusData;
      const statusColors = resolveStatusColors(options?.color);
      const processorConfig = {
        ...config,
        color: void 0,
        padding: options?.padding,
        width: options?.width,
        _statusColors: statusColors
      };
      const processor = new processor_default(processorConfig);
      const [chart, processedOptions] = processor.process(statusData);
      this.chart = chart;
      this.processedOptions = processedOptions;
      const statusFormatter = new status_default(processedOptions);
      this.asciiChart = statusFormatter.format(chart);
      return;
    }
    if (chartType === "candlestick") {
      const candleData = normalizeCandlestickInput(data);
      this.originalData = candleData;
      const { bullColor, bearColor } = resolveCandlestickColors(options?.color);
      const processorConfig = {
        ...config,
        color: void 0,
        padding: options?.padding,
        _bullColor: bullColor,
        _bearColor: bearColor
      };
      const processor = new processor_default(processorConfig);
      const [chart, processedOptions] = processor.process(candleData);
      this.chart = chart;
      this.processedOptions = processedOptions;
      const candleFormatter = new candlestick_default(processedOptions);
      this.asciiChart = candleFormatter.format(chart);
      return;
    }
    if (isPointChartType(chartType) && isMultiSeriesData(data)) {
      const rawPoints = data;
      const seriesCount = rawPoints[0]?.length ?? 0;
      const seriesData = [];
      for (let s = 0; s < seriesCount; s++) {
        seriesData.push(rawPoints.map((point) => point[s]));
      }
      this.originalData = seriesData.flat();
      const seriesColors = resolveSeriesColors(options?.color, seriesCount);
      const processorConfig = { ...config, color: void 0 };
      const processor = new processor_default(processorConfig);
      const charts = [];
      let lastOpts = processorConfig;
      for (const series of seriesData) {
        const [chart, opts] = processor.process(series);
        charts.push(chart);
        lastOpts = opts;
      }
      lastOpts._seriesColors = seriesColors;
      this.chart = charts[0] ?? /* @__PURE__ */ new Map();
      this.processedOptions = lastOpts;
      let multiFormatter;
      if (chartType === "step") {
        multiFormatter = new step_default(lastOpts);
      } else if (chartType === "scatter") {
        multiFormatter = new scatter_default(lastOpts);
      } else {
        multiFormatter = new line_default(lastOpts);
      }
      this.asciiChart = multiFormatter.formatMulti(charts);
    } else {
      const inputData = data;
      this.originalData = inputData;
      let preProcessConfig = config;
      if (isPointChartType(chartType)) {
        const isScatterAuto = chartType === "scatter" && options?.color === "auto";
        if (!isScatterAuto) {
          const seriesColors = resolveSeriesColors(options?.color, 1);
          preProcessConfig = { ...config, color: void 0, _seriesColors: seriesColors };
        }
      }
      const processor = new processor_default(preProcessConfig);
      const [chart, processedOptions] = processor.process(inputData);
      this.chart = chart;
      this.processedOptions = processedOptions;
      let chartFormatter;
      if (chartType === "step") {
        chartFormatter = new step_default(processedOptions);
      } else if (chartType === "line") {
        chartFormatter = new line_default(processedOptions);
      } else if (chartType === "scatter") {
        chartFormatter = new scatter_default(processedOptions);
      } else if (config.orientation === "vertical") {
        chartFormatter = new vertical_default(processedOptions);
      } else {
        chartFormatter = new horizontal_default(processedOptions);
      }
      this.asciiChart = chartFormatter.format(this.chart);
    }
  }
  create() {
    return this.asciiChart;
  }
  createAt(progress) {
    const clampedProgress = Math.max(0, Math.min(1, progress));
    const chartType = this.processedOptions.type;
    if (usesProgressReveal(chartType)) {
      const fixedOptions2 = {
        ...this.options,
        _animationProgress: clampedProgress
      };
      const chart2 = new _Chartscii(this.originalRawData, fixedOptions2);
      return chart2.create();
    }
    const scaledData = scaleInputData(this.originalData, clampedProgress, 2, true);
    const originalMax = getMaxValue(this.originalData);
    const size = this.processedOptions.orientation === "vertical" ? this.processedOptions.height || 20 : this.processedOptions.width || 80;
    const fixedScale = originalMax / size;
    const charLength = this.processedOptions.char?.length || 1;
    const finalMaxBarLength = Math.floor(this.processedOptions.max.scaled / charLength);
    const fixedOptions = {
      ...this.options,
      scale: fixedScale,
      width: this.processedOptions.width,
      height: this.processedOptions.height,
      _maxLabel: this.processedOptions.max.label,
      // Internal: preserve label width for animation
      _finalMaxBarLength: finalMaxBarLength
      // Internal: preserve max bar length for gradient fill
    };
    const chart = new _Chartscii(scaledData, fixedOptions);
    return chart.create();
  }
  async animate(options = {}) {
    const { duration = 1e3, fps = 30, easing = "easeOut", step, frames: returnFrames } = options;
    const easingFn = easings[easing];
    const output = this.create();
    const lineCount = output.split("\n").length;
    const collectedFrames = [];
    if (step !== void 0 && step > 0) {
      const steps = Math.ceil(1 / step);
      const frameTime = duration / steps;
      for (let i = 0; i <= steps; i++) {
        const t = i / steps;
        const progress = easingFn(t);
        const frameOutput = this.createAt(progress);
        if (returnFrames) {
          collectedFrames.push(frameOutput);
        } else {
          if (i > 0) {
            process.stdout.write(`\x1B[${lineCount}A`);
          }
          process.stdout.write(frameOutput + "\n");
          if (i < steps) {
            await new Promise((resolve) => setTimeout(resolve, frameTime));
          }
        }
      }
    } else {
      const frameTime = 1e3 / fps;
      const totalFrames = Math.ceil(duration / frameTime);
      for (let frame = 0; frame <= totalFrames; frame++) {
        const t = frame / totalFrames;
        const progress = easingFn(t);
        const frameOutput = this.createAt(progress);
        if (returnFrames) {
          collectedFrames.push(frameOutput);
        } else {
          if (frame > 0) {
            process.stdout.write(`\x1B[${lineCount}A`);
          }
          process.stdout.write(frameOutput + "\n");
          if (frame < totalFrames) {
            await new Promise((resolve) => setTimeout(resolve, frameTime));
          }
        }
      }
    }
    if (returnFrames) {
      return collectedFrames;
    }
  }
};
var chartscii_default = Chartscii;

// index.ts
var index_default = chartscii_default;
export {
  chartscii_default as Chartscii,
  index_default as default
};
