// styl3/dist/style.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const colors_1 = __importStar(require("./colors/colors"));
const decorators_1 = require("./decorators/decorators");
const utils_1 = require("./utils/utils");
function hexToRgb(hex) {
    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
    return result
        ? {
            r: parseInt(result[1], 16),
            g: parseInt(result[2], 16),
            b: parseInt(result[3], 16)
        }
        : null;
}
function rgbToAnsi(r, g, b, txt) {
    return `\x1b[38;2;${r};${g};${b}m${txt || ''}${txt ? colors_1.RESET : ''}`;
}
function makeColors(codes) {
    return Object.keys(codes).reduce((acc, code) => {
        let value = codes[String(code)];
        if (Array.isArray(value))
            value = value.join(';');
        if (typeof value === 'string' && value.toString().includes('#')) {
            const { r, g, b } = hexToRgb(value);
            value = rgbToAnsi(r, g, b);
            acc[code] = value;
            return acc;
        }
        if (typeof value === 'object') {
            acc[code] = makeColors(value);
            return acc;
        }
        acc[code] = `\x1b[${value}m`;
        return acc;
    }, {});
}
function applySymbols(symbols, value, formattedColor) {
    const keys = Object.keys(symbols);
    return keys.reduce((acc, key) => {
        const symbol = symbols[key];
        const regexString = `\\${symbol}(.*)\\` + symbol;
        const regex = new RegExp(regexString, 'gsm');
        if (acc && acc.match(regex)) {
            const [subject, stripped] = regex.exec(acc);
            const replaced = decorators_1.symbolsFunctions[key](formattedColor + stripped + formattedColor);
            acc = acc.replace(subject, replaced);
        }
        return acc;
    }, value);
}
function makeFunctions(colors, symbols) {
    return Object.keys(colors).reduce((acc, color) => {
        if (typeof colors[color] === 'object') {
            acc[color] = makeFunctions(colors[color], symbols);
            return acc;
        }
        acc[color] = function (text, ...args) {
            const formattedColor = colors[color];
            const value = typeof text === "string" ? text : (0, utils_1.getValue)(text, ...args);
            const formattedValue = applySymbols(symbols, value, formattedColor);
            return `${formattedColor}${formattedValue}${colors_1.RESET}`;
        };
        return acc;
    }, {});
}
function rgb(r, g, b) {
    return function (text, ...args) {
        if (typeof text === "string") {
            return rgbToAnsi(r, g, b, text);
        }
        else {
            const value = (0, utils_1.getValue)(text, ...args);
            return rgbToAnsi(r, g, b, value);
        }
    };
}
function hex(hex) {
    return function (text, ...args) {
        const { r, g, b } = hexToRgb(hex);
        if (typeof text === "string") {
            return rgbToAnsi(r, g, b, text);
        }
        else {
            const value = (0, utils_1.getValue)(text, ...args);
            return rgbToAnsi(r, g, b, value);
        }
    };
}
function ansi(ansi) {
    return function (text, ...args) {
        if (typeof text === "string") {
            return `${ansi}${text}${colors_1.RESET}`;
        }
        else {
            const value = (0, utils_1.getValue)(text, ...args);
            return `${ansi}${value}${colors_1.RESET}`;
        }
    };
}
function style(config) {
    const { colors, theme, decorators } = config || { decorators: decorators_1.decoratorMap, colors: colors_1.default, theme: 'default' };
    const allColors = { ...colors_1.default, ...colors };
    const allSymbols = { ...decorators_1.decoratorMap, ...decorators };
    const colorCodes = makeColors(allColors);
    const themedColors = theme ? { ...colorCodes, ...colorCodes[theme] } : colorCodes;
    if (!themedColors)
        throw new Error(`no such theme ${theme.toString()}`);
    const functions = makeFunctions(themedColors, allSymbols);
    const styled = {
        colors: themedColors,
        symbols: allSymbols,
        ...functions,
        ...decorators_1.decoratorFunctions,
        rgb,
        hex,
        ansi
    };
    return styled;
}
exports.default = style;
